{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "508e3fda",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: generics\n",
      "\n",
      "\n",
      "Attaching package: ‘generics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    explain\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
      "    setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
      "    unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:generics’:\n",
      "\n",
      "    components, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:igraph’:\n",
      "\n",
      "    compose, simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(stringr)\n",
    "library(IRanges)\n",
    "library(igraph)\n",
    "library(purrr)\n",
    "library(tidyr)\n",
    "library(parallel)\n",
    "library(pbapply)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6fd96d63",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Functions #\n",
    "\n",
    "find_motif_cluster_in_sea <- function(sea_res,mc,full.only=F,min_nmotif=3,btwn_mt_max_gap=30,ncores=4) {\n",
    "  sea_res_sort <- sea_res %>% group_by(seq_name) %>% arrange(start)\n",
    "  \n",
    "  meme_size <- sea_res[!duplicated(sea_res$meme_name),'site_width']\n",
    "  names(meme_size) <- sea_res[!duplicated(sea_res$meme_name),'meme_name']\n",
    "  sea_spl <- split(sea_res, sea_res$seq_name)\n",
    "\n",
    "  # Consecutive meme pairs in mc for consecutive gap check\n",
    "  consec_mp  <- c()\n",
    "  for (i in 1:(length(mc)-1)) {\n",
    "    consec_mp <- c(consec_mp, paste0(names(mc)[i],'_',names(mc)[i+1]))\n",
    "  }\n",
    "\n",
    "  # Force progress bar to show\n",
    "  pboptions(type = \"txt\")\n",
    "  \n",
    "  clst_found <- pblapply(sea_spl,function(x) {\n",
    "    name <- x$seq_name %>% unique\n",
    "    # message(Sys.time(),' Processing ', name, ' with ', nrow(x), ' motifs')\n",
    "    gene_strand <- unique(x$strand)\n",
    "    x_ranges <- IRanges(start = x$start, width = x$site_width, names = row.names(x))\n",
    "    hits <- findOverlaps(x_ranges, drop.self = TRUE)\n",
    "    x_nonovl_combs <- list()\n",
    "    \n",
    "    if (length(hits)>0) {\n",
    "      overlapping_vertices <- unique(c(queryHits(hits), subjectHits(hits)))\n",
    "      overlapping_motif_ids <- row.names(x)[overlapping_vertices]\n",
    "      \n",
    "      edge_list <- data.frame(\n",
    "        from = row.names(x)[queryHits(hits)],\n",
    "        to = row.names(x)[subjectHits(hits)]\n",
    "      )\n",
    "      \n",
    "      g <- graph_from_data_frame(edge_list, directed = FALSE, vertices = data.frame(name = overlapping_motif_ids))\n",
    "      components <- components(g)\n",
    "      overlapping_groups <- split(names(components$membership), components$membership)\n",
    "      \n",
    "      non_overlapping_motifs <- x[!row.names(x) %in% V(g)$name, ]\n",
    "      \n",
    "      overlapping_ids_list <- map(overlapping_groups, ~ .x)\n",
    "      combinations <- cross(overlapping_ids_list)\n",
    "      \n",
    "      # message(Sys.time(),' Processing combinations')\n",
    "      for (i in 1:length(combinations)) {\n",
    "        # message(Sys.time(), ' Processing combination ', i, '/', length(combinations))\n",
    "        current_ids <- unlist(combinations[[i]])\n",
    "        current_df <- rbind(\n",
    "          non_overlapping_motifs,\n",
    "          x[row.names(x) %in% current_ids, ]\n",
    "        )\n",
    "        current_df <- current_df[order(current_df$site_start),]\n",
    "        comb_id <- paste(current_df[current_ids,'meme_name'],current_ids,sep='_')\n",
    "        comb_id <- paste(comb_id, collapse = \"_\")\n",
    "        x_nonovl_combs[[paste0(\"Combination_\", comb_id)]] <- current_df\n",
    "      }\n",
    "    } else {\n",
    "      x_nonovl_combs[['Orig_data']] <- x[order(x$site_start),]\n",
    "    }\n",
    "    \n",
    "    clst_found <- matrix(nrow = 0,ncol=8) %>% as.data.frame()\n",
    "    colnames(clst_found) <- c('combination','seq_name','chr','strand',\n",
    "                              'meme_start_site_start','meme_end_site_start','meme_seq','ref_motif')\n",
    "    # clst_found <- list()\n",
    "    # message('Number of nonovl combinations',': ',length(x_nonovl_combs))\n",
    "    kidx=1\n",
    "    for (k in names(x_nonovl_combs)) {\n",
    "      # message('Processing nonovl combination ',kidx,'/',length(x_nonovl_combs))\n",
    "      kidx = kidx + 1\n",
    "      nonovl_comb <- x_nonovl_combs[[k]]\n",
    "      if (nrow(nonovl_comb)>=min_nmotif) {\n",
    "        widest_ws <- min(length(mc),nrow(nonovl_comb))\n",
    "        for (window_size in widest_ws:min_nmotif) {\n",
    "          windows <- list()\n",
    "          j=1\n",
    "          for (i in seq(1,nrow(nonovl_comb)-window_size+1)) {\n",
    "            slide = nonovl_comb[i:(i+window_size-1),]\n",
    "            no_duplicate = !(duplicated(slide$meme_name) %>% any())\n",
    "            if (no_duplicate){\n",
    "              windows[[j]] <- slide\n",
    "              j = j+1\n",
    "            }\n",
    "          }\n",
    "          if (length(windows) >= 1) {\n",
    "            for (i in seq(length(windows))){\n",
    "              slide = windows[[i]]\n",
    "              if (gene_strand=='+') {\n",
    "                coord_sorted <- slide[order(slide$start),]\n",
    "                coord_sorted <- coord_sorted %>% mutate(\n",
    "                                gap_to_next = ifelse(row_number() < n(), \n",
    "                                                      lead(start) - (end), NA),\n",
    "                                gap_description = ifelse(row_number() < n(),\n",
    "                                                      paste0(meme_name, \"_\", lead(meme_name)),NA))\n",
    "              } else {\n",
    "                coord_sorted <- slide[order(slide$start,decreasing = T),]\n",
    "                coord_sorted <- coord_sorted %>% mutate(\n",
    "                                gap_to_next = ifelse(row_number() < n(), \n",
    "                                                      lag(start) - (end), NA),\n",
    "                                gap_description = ifelse(row_number() < n(),\n",
    "                                                      paste0(meme_name, \"_\", lag(meme_name)),NA))\n",
    "              }\n",
    "              consec_mp_gaps <- coord_sorted[coord_sorted$gap_description %in% consec_mp,\n",
    "                                            ]$gap_to_next\n",
    "              if (any(consec_mp_gaps > btwn_mt_max_gap, na.rm = T)) {next}\n",
    "\n",
    "              meme_seq <- coord_sorted$meme_name\n",
    "              # Flip cluster motifs if strands of corresponding motifs not match\n",
    "              a <- coord_sorted$site_strand[1]\n",
    "              b <- mc[coord_sorted$meme_name[1]] %>% unname()\n",
    "              if (a != b) {\n",
    "                mc <- mc[length(mc):1]\n",
    "                mc <- ifelse(mc=='+','-','+')\n",
    "              }\n",
    "              \n",
    "              mc_order <- 1:length(mc)\n",
    "              names(mc_order) <- names(mc)\n",
    "              \n",
    "              order <- mc_order[meme_seq] %>% unname()\n",
    "              \n",
    "              order_check <- identical(order,sort(order))\n",
    "              strandness_check <- identical(coord_sorted$site_strand,unname(mc[meme_seq]))\n",
    "              \n",
    "              if (order_check & strandness_check) {\n",
    "                gaps <- data.frame(meme_start = order[which(abs(diff(order))>1)], \n",
    "                                   meme_end = order[which(abs(diff(order))>1)+1])\n",
    "                if (nrow(gaps)!=0) {\n",
    "                  gaps <- apply(gaps,1,function(x){\n",
    "                    mm <- mc_order[(x[1]+1):(x[2]-1)] %>% names()\n",
    "                    mm_size <- sum(meme_size[mm])\n",
    "                    meme_start <- mc_order[mc_order %in% x[1]] %>% names()\n",
    "                    meme_end <- mc_order[mc_order %in% x[2]] %>% names()\n",
    "                    gap_size <- coord_sorted$site_start[which(coord_sorted$meme_name == meme_end)] - \n",
    "                      (coord_sorted$site_start[which(coord_sorted$meme_name == meme_start)] + \n",
    "                         coord_sorted$site_width[which(coord_sorted$meme_name == meme_start)])\n",
    "                    mm <- paste0(mm,collapse = ',')\n",
    "                    df <- data.frame(gap_start = meme_start, gap_end = meme_end, missing_motifs = mm,\n",
    "                                     missing_motifs_size = mm_size, gap_size = gap_size)\n",
    "                    return(df)\n",
    "                  })\n",
    "                  gaps <- do.call('rbind',gaps)\n",
    "                } else {\n",
    "                  gaps <- matrix(ncol = 5, nrow = 0) %>% as.data.frame()\n",
    "                  colnames(gaps) <- c('gap_start','gap_end','missing_motifs','missing_motifs_size','gap_size')\n",
    "                }\n",
    "                \n",
    "                res <- data.frame(combination = k,\n",
    "                                  seq_name = name,\n",
    "                                  chr = unique(coord_sorted$chr),\n",
    "                                  strand = gene_strand,\n",
    "                                  meme_start_site_start = coord_sorted$site_start[1],\n",
    "                                  meme_end_site_start = coord_sorted$site_start[length(coord_sorted$start)],\n",
    "                                  meme_seq = paste0(meme_seq,collapse = ','),\n",
    "                                  ref_motif = paste0(names(mc),collapse = ','),\n",
    "                                  min_consec_gap = coord_sorted$gap_to_next %>% min(na.rm = T))\n",
    "                # if (name %in% seq_uniq_in_meme) {print(res)}\n",
    "                \n",
    "                if (nrow(gaps)==0){\n",
    "                  clst_found <- rbind(clst_found,res)\n",
    "                } else {\n",
    "                  gaps$pass <- apply(gaps,1,function(x){\n",
    "                    num_mm <- strsplit(x[3],split = ',') %>% unlist %>% length\n",
    "                    mm_size <- x[4] %>% as.numeric()\n",
    "                    max_allowed_gs <- mm_size + (num_mm+1)*btwn_mt_max_gap\n",
    "                    gs <- x[5] %>% as.numeric\n",
    "                    pass_or_not <- (mm_size < gs) & (gs < max_allowed_gs)\n",
    "                    return(pass_or_not)\n",
    "                  })\n",
    "                  # if (name %in% seq_uniq_in_meme) {print(gaps)}\n",
    "                  if (all(gaps$pass)){\n",
    "                    clst_found <- rbind(clst_found,res)\n",
    "                  }\n",
    "                }\n",
    "                \n",
    "                #  clst_found[[paste0(k,'_',l)]] <- list(res = res, gaps = gaps)\n",
    "                #  l = l + 1\n",
    "              }\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          if (nrow(clst_found)>0){break}\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    return(clst_found)\n",
    "  }, cl = ncores)\n",
    "  return(clst_found)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4c3418c1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "wd <- \"/Users/ninhle/Desktop/Research/scPASU_pipeline_runs/Ureter10_scPASU_run/outputs/differentiation_stage_cellranger_peakcount/\"\n",
    "setwd(wd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c1809c8a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>seq_name</th><th scope=col>score</th><th scope=col>strand</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>  70006</td><td>  71586</td><td>TU4_UTR1 </td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td> 944151</td><td> 944576</td><td>TU17_UTR2</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td> 965189</td><td> 965720</td><td>TU18_UTR3</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td> 974573</td><td> 975866</td><td>TU19_UTR4</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>1014476</td><td>1014541</td><td>TU20_UTR5</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>1054979</td><td>1056119</td><td>TU21_UTR6</td><td>1</td><td>+</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & chr & start & end & seq\\_name & score & strand\\\\\n",
       "  & <chr> & <int> & <int> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 &   70006 &   71586 & TU4\\_UTR1  & 1 & +\\\\\n",
       "\t2 & chr1 &  944151 &  944576 & TU17\\_UTR2 & 1 & +\\\\\n",
       "\t3 & chr1 &  965189 &  965720 & TU18\\_UTR3 & 1 & +\\\\\n",
       "\t4 & chr1 &  974573 &  975866 & TU19\\_UTR4 & 1 & +\\\\\n",
       "\t5 & chr1 & 1014476 & 1014541 & TU20\\_UTR5 & 1 & +\\\\\n",
       "\t6 & chr1 & 1054979 & 1056119 & TU21\\_UTR6 & 1 & +\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | chr &lt;chr&gt; | start &lt;int&gt; | end &lt;int&gt; | seq_name &lt;chr&gt; | score &lt;int&gt; | strand &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 |   70006 |   71586 | TU4_UTR1  | 1 | + |\n",
       "| 2 | chr1 |  944151 |  944576 | TU17_UTR2 | 1 | + |\n",
       "| 3 | chr1 |  965189 |  965720 | TU18_UTR3 | 1 | + |\n",
       "| 4 | chr1 |  974573 |  975866 | TU19_UTR4 | 1 | + |\n",
       "| 5 | chr1 | 1014476 | 1014541 | TU20_UTR5 | 1 | + |\n",
       "| 6 | chr1 | 1054979 | 1056119 | TU21_UTR6 | 1 | + |\n",
       "\n"
      ],
      "text/plain": [
       "  chr  start   end     seq_name  score strand\n",
       "1 chr1   70006   71586 TU4_UTR1  1     +     \n",
       "2 chr1  944151  944576 TU17_UTR2 1     +     \n",
       "3 chr1  965189  965720 TU18_UTR3 1     +     \n",
       "4 chr1  974573  975866 TU19_UTR4 1     +     \n",
       "5 chr1 1014476 1014541 TU20_UTR5 1     +     \n",
       "6 chr1 1054979 1056119 TU21_UTR6 1     +     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Variables\n",
    "fprefix <- 'all_TU_disjoint_utr3'\n",
    "# fprefix <- 'all_TU_disjoint_utr3_5kbflank'\n",
    "\n",
    "seq_bed <- read.delim(paste0(fprefix,'.bed'),\n",
    "                      header = F)\n",
    "colnames(seq_bed) <- c('chr','start','end','seq_name','score','strand')\n",
    "head(seq_bed)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56a560b5",
   "metadata": {},
   "source": [
    "##### Lengthening motif cluster detecting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c85ff2f3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running sea cluster detection for sea_out_lgthn_all_TU_disjoint_utr3...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 10 × 2\u001b[39m\n",
      "   seq_name             n\n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m 1\u001b[39m TU22647_UTR12934    88\n",
      "\u001b[90m 2\u001b[39m TU4659_UTR23842     79\n",
      "\u001b[90m 3\u001b[39m TU26609_UTR15674    75\n",
      "\u001b[90m 4\u001b[39m TU32685_UTR20212    69\n",
      "\u001b[90m 5\u001b[39m TU26145_UTR15390    63\n",
      "\u001b[90m 6\u001b[39m TU36069_UTR25042    63\n",
      "\u001b[90m 7\u001b[39m TU23086_UTR13233    59\n",
      "\u001b[90m 8\u001b[39m TU3786_UTR23194     59\n",
      "\u001b[90m 9\u001b[39m TU13708_UTR5877     56\n",
      "\u001b[90m10\u001b[39m TU418_UTR348        52\n",
      "  |++++++++++++++++++++++++++++++++++++++++++++++++++| 100%\n"
     ]
    }
   ],
   "source": [
    "#### SEA\n",
    "filedir=paste0('sea_out_lgthn_',fprefix)\n",
    "message(paste0('Running sea cluster detection for ',filedir,'...'))\n",
    "  \n",
    "sea_res <- read.delim(paste0(filedir,'/sites.tsv'))\n",
    "sea_res <- sea_res[!sea_res$motif_ID=='#',]\n",
    "sea_res$strand <- str_extract_all(sea_res$seq_ID, \"\\\\(([^)]+)\\\\)\") %>% unlist\n",
    "sea_res$strand <- gsub(\"\\\\(|\\\\)\",\"\",sea_res$strand)\n",
    "sea_res$cis <- sea_res$site_Strand==sea_res$strand\n",
    "sea_res$seq_ID <- gsub(\"\\\\(([^)]+)\\\\)\",\"\",sea_res$seq_ID)\n",
    "sea_res$chr <- seq_bed$chr[match(sea_res$seq_ID,seq_bed$seq_name)]\n",
    "sea_res$range_start <- seq_bed$start[match(sea_res$seq_ID,seq_bed$seq_name)]\n",
    "sea_res$range_end <- seq_bed$end[match(sea_res$seq_ID,seq_bed$seq_name)]\n",
    "sea_res$site_width <- nchar(sea_res$site_Sequence)\n",
    "sea_res$start <- ifelse(sea_res$strand=='+',sea_res$range_start+sea_res$site_Start-1,\n",
    "                                                sea_res$range_end-sea_res$site_Start-sea_res$site_width+1)\n",
    "sea_res$end <- sea_res$start+sea_res$site_width\n",
    "colnames(sea_res)[c(2,3,4,5,6,8)] <- c('meme_name','seq_name','site_start','site_end','site_strand','site')\n",
    "sea_res <- sea_res %>% select(seq_name,site_strand,site_start,site,meme_name,strand,cis,chr,\n",
    "                                                      range_start,range_end,site_width,start,end)\n",
    "sea_res <- sea_res %>% mutate(pval = NA,flank1 = NA,flank2 = NA) %>%\n",
    "    relocate(pval, .after = site_start) %>%\n",
    "    relocate(flank1, .after = pval) %>%\n",
    "    relocate(flank2, .after = site)\n",
    "top_nmotif <- sea_res %>% group_by(seq_name) %>% summarize(n=n()) %>% arrange(desc(n))\n",
    "print(top_nmotif[1:10,])\n",
    "write.table(sea_res,paste0(filedir,'/sea_res_processed_',fprefix,'.tsv'),\n",
    "            sep = '\\t',quote = F,row.names = F)                       \n",
    "sea_clst_found <- find_motif_cluster_in_sea(sea_res,mc = c('MEME-3'='+','MEME-1'='+','MEME-2'='+','MEME-5'='-','MEME-4'='+'),\n",
    "                                              min_nmotif=3,btwn_mt_max_gap=40,ncores = 1)\n",
    "saveRDS(sea_clst_found,paste0(filedir,'/sea_clst_found_',fprefix,'.rds'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d53f2483",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of UTRs with motif cluster found: 4016 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>combination</th><th scope=col>seq_name</th><th scope=col>chr</th><th scope=col>strand</th><th scope=col>meme_start_site_start</th><th scope=col>meme_end_site_start</th><th scope=col>meme_seq</th><th scope=col>ref_motif</th><th scope=col>min_consec_gap</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>TU1525_UTR1293</th><td>Orig_data                                         </td><td>TU1525_UTR1293  </td><td>chr1 </td><td>+</td><td>2973</td><td>3125</td><td>MEME-1,MEME-2,MEME-5,MEME-4       </td><td>MEME-3,MEME-1,MEME-2,MEME-5,MEME-4</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU22463_UTR12756.1</th><td>Combination_MEME-2_16931                          </td><td>TU22463_UTR12756</td><td>chr12</td><td>+</td><td>1011</td><td>1270</td><td>MEME-3,MEME-5,MEME-4              </td><td>MEME-3,MEME-1,MEME-2,MEME-5,MEME-4</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU33615_UTR21115.1</th><td>Combination_MEME-2_16248                          </td><td>TU33615_UTR21115</td><td>chr19</td><td>-</td><td>8931</td><td>9084</td><td>MEME-1,MEME-2,MEME-5,MEME-4       </td><td>MEME-3,MEME-1,MEME-2,MEME-5,MEME-4</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU37452_UTR26142.3</th><td>Combination_MEME-2_12293_MEME-2_16043_MEME-2_18298</td><td>TU37452_UTR26142</td><td>chrX </td><td>+</td><td> 861</td><td>1027</td><td>MEME-4,MEME-5,MEME-2,MEME-1       </td><td>MEME-4,MEME-5,MEME-2,MEME-1,MEME-3</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU37452_UTR26142.5</th><td>Combination_MEME-2_12293_MEME-3_20713_MEME-2_18298</td><td>TU37452_UTR26142</td><td>chrX </td><td>+</td><td> 861</td><td>1113</td><td>MEME-4,MEME-5,MEME-2,MEME-1,MEME-3</td><td>MEME-4,MEME-5,MEME-2,MEME-1,MEME-3</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU37452_UTR26142.7</th><td>Combination_MEME-3_21327_MEME-3_20713_MEME-2_18298</td><td>TU37452_UTR26142</td><td>chrX </td><td>+</td><td> 861</td><td>1113</td><td>MEME-4,MEME-5,MEME-2,MEME-1,MEME-3</td><td>MEME-4,MEME-5,MEME-2,MEME-1,MEME-3</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & combination & seq\\_name & chr & strand & meme\\_start\\_site\\_start & meme\\_end\\_site\\_start & meme\\_seq & ref\\_motif & min\\_consec\\_gap\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tTU1525\\_UTR1293 & Orig\\_data                                          & TU1525\\_UTR1293   & chr1  & + & 2973 & 3125 & MEME-1,MEME-2,MEME-5,MEME-4        & MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 & 0\\\\\n",
       "\tTU22463\\_UTR12756.1 & Combination\\_MEME-2\\_16931                           & TU22463\\_UTR12756 & chr12 & + & 1011 & 1270 & MEME-3,MEME-5,MEME-4               & MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 & 0\\\\\n",
       "\tTU33615\\_UTR21115.1 & Combination\\_MEME-2\\_16248                           & TU33615\\_UTR21115 & chr19 & - & 8931 & 9084 & MEME-1,MEME-2,MEME-5,MEME-4        & MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 & 0\\\\\n",
       "\tTU37452\\_UTR26142.3 & Combination\\_MEME-2\\_12293\\_MEME-2\\_16043\\_MEME-2\\_18298 & TU37452\\_UTR26142 & chrX  & + &  861 & 1027 & MEME-4,MEME-5,MEME-2,MEME-1        & MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 & 0\\\\\n",
       "\tTU37452\\_UTR26142.5 & Combination\\_MEME-2\\_12293\\_MEME-3\\_20713\\_MEME-2\\_18298 & TU37452\\_UTR26142 & chrX  & + &  861 & 1113 & MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 & MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 & 0\\\\\n",
       "\tTU37452\\_UTR26142.7 & Combination\\_MEME-3\\_21327\\_MEME-3\\_20713\\_MEME-2\\_18298 & TU37452\\_UTR26142 & chrX  & + &  861 & 1113 & MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 & MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | combination &lt;chr&gt; | seq_name &lt;chr&gt; | chr &lt;chr&gt; | strand &lt;chr&gt; | meme_start_site_start &lt;int&gt; | meme_end_site_start &lt;int&gt; | meme_seq &lt;chr&gt; | ref_motif &lt;chr&gt; | min_consec_gap &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| TU1525_UTR1293 | Orig_data                                          | TU1525_UTR1293   | chr1  | + | 2973 | 3125 | MEME-1,MEME-2,MEME-5,MEME-4        | MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 | 0 |\n",
       "| TU22463_UTR12756.1 | Combination_MEME-2_16931                           | TU22463_UTR12756 | chr12 | + | 1011 | 1270 | MEME-3,MEME-5,MEME-4               | MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 | 0 |\n",
       "| TU33615_UTR21115.1 | Combination_MEME-2_16248                           | TU33615_UTR21115 | chr19 | - | 8931 | 9084 | MEME-1,MEME-2,MEME-5,MEME-4        | MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 | 0 |\n",
       "| TU37452_UTR26142.3 | Combination_MEME-2_12293_MEME-2_16043_MEME-2_18298 | TU37452_UTR26142 | chrX  | + |  861 | 1027 | MEME-4,MEME-5,MEME-2,MEME-1        | MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 | 0 |\n",
       "| TU37452_UTR26142.5 | Combination_MEME-2_12293_MEME-3_20713_MEME-2_18298 | TU37452_UTR26142 | chrX  | + |  861 | 1113 | MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 | MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 | 0 |\n",
       "| TU37452_UTR26142.7 | Combination_MEME-3_21327_MEME-3_20713_MEME-2_18298 | TU37452_UTR26142 | chrX  | + |  861 | 1113 | MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 | MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "                   combination                                       \n",
       "TU1525_UTR1293     Orig_data                                         \n",
       "TU22463_UTR12756.1 Combination_MEME-2_16931                          \n",
       "TU33615_UTR21115.1 Combination_MEME-2_16248                          \n",
       "TU37452_UTR26142.3 Combination_MEME-2_12293_MEME-2_16043_MEME-2_18298\n",
       "TU37452_UTR26142.5 Combination_MEME-2_12293_MEME-3_20713_MEME-2_18298\n",
       "TU37452_UTR26142.7 Combination_MEME-3_21327_MEME-3_20713_MEME-2_18298\n",
       "                   seq_name         chr   strand meme_start_site_start\n",
       "TU1525_UTR1293     TU1525_UTR1293   chr1  +      2973                 \n",
       "TU22463_UTR12756.1 TU22463_UTR12756 chr12 +      1011                 \n",
       "TU33615_UTR21115.1 TU33615_UTR21115 chr19 -      8931                 \n",
       "TU37452_UTR26142.3 TU37452_UTR26142 chrX  +       861                 \n",
       "TU37452_UTR26142.5 TU37452_UTR26142 chrX  +       861                 \n",
       "TU37452_UTR26142.7 TU37452_UTR26142 chrX  +       861                 \n",
       "                   meme_end_site_start meme_seq                          \n",
       "TU1525_UTR1293     3125                MEME-1,MEME-2,MEME-5,MEME-4       \n",
       "TU22463_UTR12756.1 1270                MEME-3,MEME-5,MEME-4              \n",
       "TU33615_UTR21115.1 9084                MEME-1,MEME-2,MEME-5,MEME-4       \n",
       "TU37452_UTR26142.3 1027                MEME-4,MEME-5,MEME-2,MEME-1       \n",
       "TU37452_UTR26142.5 1113                MEME-4,MEME-5,MEME-2,MEME-1,MEME-3\n",
       "TU37452_UTR26142.7 1113                MEME-4,MEME-5,MEME-2,MEME-1,MEME-3\n",
       "                   ref_motif                          min_consec_gap\n",
       "TU1525_UTR1293     MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 0             \n",
       "TU22463_UTR12756.1 MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 0             \n",
       "TU33615_UTR21115.1 MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 0             \n",
       "TU37452_UTR26142.3 MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 0             \n",
       "TU37452_UTR26142.5 MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 0             \n",
       "TU37452_UTR26142.7 MEME-4,MEME-5,MEME-2,MEME-1,MEME-3 0             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filedir=paste0('sea_out_lgthn_',fprefix)\n",
    "sea_clst_found <- readRDS(file = paste0(filedir,'/sea_clst_found_',fprefix,'.rds'))\n",
    "sea_clst_found_df <- do.call('rbind',sea_clst_found)\n",
    "cat(\"Number of UTRs with motif cluster found:\", length(unique(sea_clst_found_df$seq_name)), \"\\n\")\n",
    "head(sea_clst_found_df[order(sea_clst_found_df$min_consec_gap),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f1242a4f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sea_clst_found_genes <- unique(gsub('_UTR[0-9]+','',sea_clst_found_df$seq_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0bdd264c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TUs with full or partial motif cluster from discover set:  30"
     ]
    }
   ],
   "source": [
    "turef_lgthn_mc_genes <- read.delim('turef_lgthn_mc_genes.txt')\n",
    "cat('Number of TUs with full or partial motif cluster from discover set: ',length(unique(turef_lgthn_mc_genes$tu)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "221ae0ca",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TU12003'</li><li>'TU14133'</li><li>'TU17295'</li><li>'TU17713'</li><li>'TU18776'</li><li>'TU19555'</li><li>'TU25707'</li><li>'TU30807'</li><li>'TU36726'</li><li>'TU36855'</li><li>'TU8157'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TU12003'\n",
       "\\item 'TU14133'\n",
       "\\item 'TU17295'\n",
       "\\item 'TU17713'\n",
       "\\item 'TU18776'\n",
       "\\item 'TU19555'\n",
       "\\item 'TU25707'\n",
       "\\item 'TU30807'\n",
       "\\item 'TU36726'\n",
       "\\item 'TU36855'\n",
       "\\item 'TU8157'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TU12003'\n",
       "2. 'TU14133'\n",
       "3. 'TU17295'\n",
       "4. 'TU17713'\n",
       "5. 'TU18776'\n",
       "6. 'TU19555'\n",
       "7. 'TU25707'\n",
       "8. 'TU30807'\n",
       "9. 'TU36726'\n",
       "10. 'TU36855'\n",
       "11. 'TU8157'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"TU12003\" \"TU14133\" \"TU17295\" \"TU17713\" \"TU18776\" \"TU19555\" \"TU25707\"\n",
       " [8] \"TU30807\" \"TU36726\" \"TU36855\" \"TU8157\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(unique(turef_lgthn_mc_genes$tu), sea_clst_found_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56e51cc8",
   "metadata": {},
   "source": [
    "Full or partial motif cluster false negatives Lengthening: 'TU12003', 'TU14133', 'TU17295', 'TU17713', 'TU18776', 'TU19555', 'TU25707', 'TU30807', 'TU36726', 'TU36855', 'TU8157' \n",
    "\n",
    "Full or partial motif cluster fall negatives because cluster first detected in flanks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab17d077",
   "metadata": {},
   "source": [
    "##### Lengthening motif cluster cleaning up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b987e95f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "calculate_overlap <- function(start1, end1, start2, end2) {\n",
    "  overlap_start <- max(start1, start2)\n",
    "  overlap_end <- min(end1, end2)\n",
    "  overlap_length <- max(0, overlap_end - overlap_start + 1)\n",
    "  \n",
    "  range1_length <- end1 - start1 + 1\n",
    "  range2_length <- end2 - start2 + 1\n",
    "  \n",
    "  # Calculate overlap as percentage of the smaller range\n",
    "  min_length <- min(range1_length, range2_length)\n",
    "  overlap_pct <- overlap_length / min_length\n",
    "  \n",
    "  return(overlap_pct)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ce6f426d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required namespace: GenomicRanges\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>tu</th><th scope=col>gene.id</th><th scope=col>name</th><th scope=col>utr</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>  70006</td><td>  71585</td><td>1580</td><td>+</td><td>TU4 </td><td>ENSG00000186092</td><td>OR4F5  </td><td>UTR1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td> 944151</td><td> 944575</td><td> 425</td><td>+</td><td>TU17</td><td>ENSG00000187634</td><td>SAMD11 </td><td>UTR2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td> 965189</td><td> 965719</td><td> 531</td><td>+</td><td>TU18</td><td>ENSG00000187961</td><td>KLHL17 </td><td>UTR3</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td> 974573</td><td> 975865</td><td>1293</td><td>+</td><td>TU19</td><td>ENSG00000187583</td><td>PLEKHN1</td><td>UTR4</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>1014476</td><td>1014540</td><td>  65</td><td>+</td><td>TU20</td><td>ENSG00000187608</td><td>ISG15  </td><td>UTR5</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>1054979</td><td>1056118</td><td>1140</td><td>+</td><td>TU21</td><td>ENSG00000188157</td><td>AGRN   </td><td>UTR6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & seqnames & start & end & width & strand & tu & gene.id & name & utr\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 &   70006 &   71585 & 1580 & + & TU4  & ENSG00000186092 & OR4F5   & UTR1\\\\\n",
       "\t2 & chr1 &  944151 &  944575 &  425 & + & TU17 & ENSG00000187634 & SAMD11  & UTR2\\\\\n",
       "\t3 & chr1 &  965189 &  965719 &  531 & + & TU18 & ENSG00000187961 & KLHL17  & UTR3\\\\\n",
       "\t4 & chr1 &  974573 &  975865 & 1293 & + & TU19 & ENSG00000187583 & PLEKHN1 & UTR4\\\\\n",
       "\t5 & chr1 & 1014476 & 1014540 &   65 & + & TU20 & ENSG00000187608 & ISG15   & UTR5\\\\\n",
       "\t6 & chr1 & 1054979 & 1056118 & 1140 & + & TU21 & ENSG00000188157 & AGRN    & UTR6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | tu &lt;chr&gt; | gene.id &lt;chr&gt; | name &lt;chr&gt; | utr &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 |   70006 |   71585 | 1580 | + | TU4  | ENSG00000186092 | OR4F5   | UTR1 |\n",
       "| 2 | chr1 |  944151 |  944575 |  425 | + | TU17 | ENSG00000187634 | SAMD11  | UTR2 |\n",
       "| 3 | chr1 |  965189 |  965719 |  531 | + | TU18 | ENSG00000187961 | KLHL17  | UTR3 |\n",
       "| 4 | chr1 |  974573 |  975865 | 1293 | + | TU19 | ENSG00000187583 | PLEKHN1 | UTR4 |\n",
       "| 5 | chr1 | 1014476 | 1014540 |   65 | + | TU20 | ENSG00000187608 | ISG15   | UTR5 |\n",
       "| 6 | chr1 | 1054979 | 1056118 | 1140 | + | TU21 | ENSG00000188157 | AGRN    | UTR6 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start   end     width strand tu   gene.id         name    utr \n",
       "1 chr1       70006   71585 1580  +      TU4  ENSG00000186092 OR4F5   UTR1\n",
       "2 chr1      944151  944575  425  +      TU17 ENSG00000187634 SAMD11  UTR2\n",
       "3 chr1      965189  965719  531  +      TU18 ENSG00000187961 KLHL17  UTR3\n",
       "4 chr1      974573  975865 1293  +      TU19 ENSG00000187583 PLEKHN1 UTR4\n",
       "5 chr1     1014476 1014540   65  +      TU20 ENSG00000187608 ISG15   UTR5\n",
       "6 chr1     1054979 1056118 1140  +      TU21 ENSG00000188157 AGRN    UTR6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes <- readRDS(\"/Users/ninhle/Desktop/Research/mcast_analyses/genes.rds\")\n",
    "ref <- genes$utr3 %>% as.data.frame()\n",
    "head(ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "09dadcba",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 16</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seq_name</th><th scope=col>site_strand</th><th scope=col>site_start</th><th scope=col>pval</th><th scope=col>flank1</th><th scope=col>site</th><th scope=col>flank2</th><th scope=col>meme_name</th><th scope=col>strand</th><th scope=col>cis</th><th scope=col>chr</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>site_width</th><th scope=col>start</th><th scope=col>end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>TU13400_UTR5644 </td><td>+</td><td> 653</td><td>NA</td><td>NA</td><td>TGCAGTGGCGCGATCTCGGCTCACTGCAACCT</td><td>NA</td><td>MEME-5</td><td>-</td><td>FALSE</td><td>chr6 </td><td>116274858</td><td>116278520</td><td>32</td><td>116277836</td><td>116277868</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>TU17531_UTR8696 </td><td>+</td><td>1803</td><td>NA</td><td>NA</td><td>TGCAGTGGCGCGATCTCGGCTCACTGCAACCT</td><td>NA</td><td>MEME-5</td><td>+</td><td> TRUE</td><td>chr9 </td><td>124810410</td><td>124814886</td><td>32</td><td>124812212</td><td>124812244</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>TU15342_UTR7163 </td><td>+</td><td> 303</td><td>NA</td><td>NA</td><td>TGCAGTGGCGCGATCTCGGCTCACTGCAACCT</td><td>NA</td><td>MEME-5</td><td>-</td><td>FALSE</td><td>chr7 </td><td>139043515</td><td>139047597</td><td>32</td><td>139047263</td><td>139047295</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>TU24917_UTR14518</td><td>-</td><td> 640</td><td>NA</td><td>NA</td><td>TGCAGTGGCGCGATCTCGGCTCACTGCAACCT</td><td>NA</td><td>MEME-5</td><td>-</td><td> TRUE</td><td>chr13</td><td>113324845</td><td>113325550</td><td>32</td><td>113324879</td><td>113324911</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>TU23065_UTR13222</td><td>+</td><td>1965</td><td>NA</td><td>NA</td><td>TGCAGTGGCGCGATCTCGGCTCACTGCAACCT</td><td>NA</td><td>MEME-5</td><td>+</td><td> TRUE</td><td>chr12</td><td>131943584</td><td>131945897</td><td>32</td><td>131945548</td><td>131945580</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>TU8804_UTR28907 </td><td>-</td><td>3429</td><td>NA</td><td>NA</td><td>TGCAGTGGCGCGATCTCGGCTCACTGCAACCT</td><td>NA</td><td>MEME-5</td><td>+</td><td>FALSE</td><td>chr4 </td><td>127832872</td><td>127840734</td><td>32</td><td>127836300</td><td>127836332</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 16\n",
       "\\begin{tabular}{r|llllllllllllllll}\n",
       "  & seq\\_name & site\\_strand & site\\_start & pval & flank1 & site & flank2 & meme\\_name & strand & cis & chr & range\\_start & range\\_end & site\\_width & start & end\\\\\n",
       "  & <chr> & <chr> & <int> & <lgl> & <lgl> & <chr> & <lgl> & <chr> & <chr> & <lgl> & <chr> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & TU13400\\_UTR5644  & + &  653 & NA & NA & TGCAGTGGCGCGATCTCGGCTCACTGCAACCT & NA & MEME-5 & - & FALSE & chr6  & 116274858 & 116278520 & 32 & 116277836 & 116277868\\\\\n",
       "\t2 & TU17531\\_UTR8696  & + & 1803 & NA & NA & TGCAGTGGCGCGATCTCGGCTCACTGCAACCT & NA & MEME-5 & + &  TRUE & chr9  & 124810410 & 124814886 & 32 & 124812212 & 124812244\\\\\n",
       "\t3 & TU15342\\_UTR7163  & + &  303 & NA & NA & TGCAGTGGCGCGATCTCGGCTCACTGCAACCT & NA & MEME-5 & - & FALSE & chr7  & 139043515 & 139047597 & 32 & 139047263 & 139047295\\\\\n",
       "\t4 & TU24917\\_UTR14518 & - &  640 & NA & NA & TGCAGTGGCGCGATCTCGGCTCACTGCAACCT & NA & MEME-5 & - &  TRUE & chr13 & 113324845 & 113325550 & 32 & 113324879 & 113324911\\\\\n",
       "\t5 & TU23065\\_UTR13222 & + & 1965 & NA & NA & TGCAGTGGCGCGATCTCGGCTCACTGCAACCT & NA & MEME-5 & + &  TRUE & chr12 & 131943584 & 131945897 & 32 & 131945548 & 131945580\\\\\n",
       "\t6 & TU8804\\_UTR28907  & - & 3429 & NA & NA & TGCAGTGGCGCGATCTCGGCTCACTGCAACCT & NA & MEME-5 & + & FALSE & chr4  & 127832872 & 127840734 & 32 & 127836300 & 127836332\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 16\n",
       "\n",
       "| <!--/--> | seq_name &lt;chr&gt; | site_strand &lt;chr&gt; | site_start &lt;int&gt; | pval &lt;lgl&gt; | flank1 &lt;lgl&gt; | site &lt;chr&gt; | flank2 &lt;lgl&gt; | meme_name &lt;chr&gt; | strand &lt;chr&gt; | cis &lt;lgl&gt; | chr &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | site_width &lt;int&gt; | start &lt;int&gt; | end &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | TU13400_UTR5644  | + |  653 | NA | NA | TGCAGTGGCGCGATCTCGGCTCACTGCAACCT | NA | MEME-5 | - | FALSE | chr6  | 116274858 | 116278520 | 32 | 116277836 | 116277868 |\n",
       "| 2 | TU17531_UTR8696  | + | 1803 | NA | NA | TGCAGTGGCGCGATCTCGGCTCACTGCAACCT | NA | MEME-5 | + |  TRUE | chr9  | 124810410 | 124814886 | 32 | 124812212 | 124812244 |\n",
       "| 3 | TU15342_UTR7163  | + |  303 | NA | NA | TGCAGTGGCGCGATCTCGGCTCACTGCAACCT | NA | MEME-5 | - | FALSE | chr7  | 139043515 | 139047597 | 32 | 139047263 | 139047295 |\n",
       "| 4 | TU24917_UTR14518 | - |  640 | NA | NA | TGCAGTGGCGCGATCTCGGCTCACTGCAACCT | NA | MEME-5 | - |  TRUE | chr13 | 113324845 | 113325550 | 32 | 113324879 | 113324911 |\n",
       "| 5 | TU23065_UTR13222 | + | 1965 | NA | NA | TGCAGTGGCGCGATCTCGGCTCACTGCAACCT | NA | MEME-5 | + |  TRUE | chr12 | 131943584 | 131945897 | 32 | 131945548 | 131945580 |\n",
       "| 6 | TU8804_UTR28907  | - | 3429 | NA | NA | TGCAGTGGCGCGATCTCGGCTCACTGCAACCT | NA | MEME-5 | + | FALSE | chr4  | 127832872 | 127840734 | 32 | 127836300 | 127836332 |\n",
       "\n"
      ],
      "text/plain": [
       "  seq_name         site_strand site_start pval flank1\n",
       "1 TU13400_UTR5644  +            653       NA   NA    \n",
       "2 TU17531_UTR8696  +           1803       NA   NA    \n",
       "3 TU15342_UTR7163  +            303       NA   NA    \n",
       "4 TU24917_UTR14518 -            640       NA   NA    \n",
       "5 TU23065_UTR13222 +           1965       NA   NA    \n",
       "6 TU8804_UTR28907  -           3429       NA   NA    \n",
       "  site                             flank2 meme_name strand cis   chr  \n",
       "1 TGCAGTGGCGCGATCTCGGCTCACTGCAACCT NA     MEME-5    -      FALSE chr6 \n",
       "2 TGCAGTGGCGCGATCTCGGCTCACTGCAACCT NA     MEME-5    +       TRUE chr9 \n",
       "3 TGCAGTGGCGCGATCTCGGCTCACTGCAACCT NA     MEME-5    -      FALSE chr7 \n",
       "4 TGCAGTGGCGCGATCTCGGCTCACTGCAACCT NA     MEME-5    -       TRUE chr13\n",
       "5 TGCAGTGGCGCGATCTCGGCTCACTGCAACCT NA     MEME-5    +       TRUE chr12\n",
       "6 TGCAGTGGCGCGATCTCGGCTCACTGCAACCT NA     MEME-5    +      FALSE chr4 \n",
       "  range_start range_end site_width start     end      \n",
       "1 116274858   116278520 32         116277836 116277868\n",
       "2 124810410   124814886 32         124812212 124812244\n",
       "3 139043515   139047597 32         139047263 139047295\n",
       "4 113324845   113325550 32         113324879 113324911\n",
       "5 131943584   131945897 32         131945548 131945580\n",
       "6 127832872   127840734 32         127836300 127836332"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Group overlapping ranges within each seq_name\n",
    "filedir=paste0('sea_out_lgthn_',fprefix)\n",
    "sea_res <- read.delim(paste0(filedir,'/sea_res_processed_',fprefix,'.tsv'))\n",
    "head(sea_res)\n",
    "sea_clst_found <- readRDS(file = paste0(filedir,'/sea_clst_found_',fprefix,'.rds'))\n",
    "sea_clst_found_df <- do.call('rbind',sea_clst_found)\n",
    "sea_clst_grouped <- sea_clst_found_df %>%\n",
    "  group_by(seq_name) %>%\n",
    "  arrange(meme_start_site_start) %>%\n",
    "  mutate(\n",
    "    range_length = meme_end_site_start - meme_start_site_start + 1,\n",
    "    overlap_group = {\n",
    "      n_rows <- n()\n",
    "      if (n_rows == 1) {\n",
    "        1\n",
    "      } else {\n",
    "        groups <- rep(NA, n_rows)\n",
    "        groups[1] <- 1\n",
    "        current_group <- 1\n",
    "        \n",
    "        for (i in 2:n_rows) {\n",
    "          # Check overlap with previous rows in current group\n",
    "          overlaps_with_group <- FALSE\n",
    "          for (j in 1:(i-1)) {\n",
    "            if (groups[j] == current_group) {\n",
    "              overlap_pct <- calculate_overlap(\n",
    "                meme_start_site_start[i], meme_end_site_start[i],\n",
    "                meme_start_site_start[j], meme_end_site_start[j]\n",
    "              )\n",
    "              if (overlap_pct >= 0.9) {\n",
    "                overlaps_with_group <- TRUE\n",
    "                break\n",
    "              }\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          if (overlaps_with_group) {\n",
    "            groups[i] <- current_group\n",
    "          } else {\n",
    "            current_group <- current_group + 1\n",
    "            groups[i] <- current_group\n",
    "          }\n",
    "        }\n",
    "        groups\n",
    "      }\n",
    "    }\n",
    "  ) %>%\n",
    "  group_by(seq_name, overlap_group) %>%\n",
    "  summarize(\n",
    "    cluster_count = n(),\n",
    "    all_combinations = paste(unique(combination), collapse = \";\"),\n",
    "    all_meme_seqs = paste(unique(meme_seq), collapse = \";\"),\n",
    "    all_starts = {\n",
    "    unique_seqs <- unique(meme_seq)\n",
    "    start_values <- sapply(unique_seqs, function(seq) meme_start_site_start[meme_seq == seq][1])\n",
    "    list(start_values)},\n",
    "    all_ends = {\n",
    "    unique_seqs <- unique(meme_seq)\n",
    "    end_values <- sapply(unique_seqs, function(seq) meme_end_site_start[meme_seq == seq][1])\n",
    "    list(end_values)},\n",
    "    .groups = \"drop\"\n",
    "  ) %>%\n",
    "  rowwise() %>%\n",
    "  mutate(\n",
    "    # Count motifs in each meme_seq\n",
    "    meme_seqs_list = list(strsplit(all_meme_seqs, \";\")[[1]]),\n",
    "    motif_counts = list(sapply(meme_seqs_list, function(x) length(strsplit(x, \",\")[[1]]))),\n",
    "    max_motif_count = max(unlist(motif_counts)),\n",
    "    \n",
    "    # Find meme_seqs with maximum motif count\n",
    "    max_meme_seqs = list(meme_seqs_list[motif_counts == max_motif_count]),\n",
    "  \n",
    "    # Extract final values directly\n",
    "    meme_seqs = if(length(max_meme_seqs) == 1) {\n",
    "    max_meme_seqs\n",
    "    } else {\n",
    "      paste(max_meme_seqs, collapse = \";\")\n",
    "    },\n",
    "    full_motif_cluster = max_motif_count == 5,\n",
    "\n",
    "    min_start = if(length(max_meme_seqs) == 1) {\n",
    "      max_seq = max_meme_seqs\n",
    "      seq_idx = which(meme_seqs_list == max_seq)\n",
    "      all_starts[seq_idx]\n",
    "    } else {\n",
    "      max_indices = which(motif_counts == max_motif_count)\n",
    "      min(all_starts[max_indices])},\n",
    "    \n",
    "    max_end = if(length(max_meme_seqs) == 1) {\n",
    "      max_seq = max_meme_seqs\n",
    "      seq_idx = which(meme_seqs_list == max_seq)\n",
    "      all_ends[seq_idx]\n",
    "    } else {\n",
    "      max_indices = which(motif_counts == max_motif_count)\n",
    "      max(all_ends[max_indices])\n",
    "    },\n",
    "  \n",
    "  combinations = all_combinations\n",
    ") %>%\n",
    "  ungroup() %>%\n",
    "  select(seq_name, overlap_group, cluster_count, full_motif_cluster, min_start, max_end, combinations, meme_seqs) %>%\n",
    "  arrange(seq_name, min_start)\n",
    "sea_clst_grouped$tu <- gsub('_UTR[0-9]+','',sea_clst_grouped$seq_name)\n",
    "sea_clst_grouped$gene <- ref$name[match(sea_clst_grouped$tu,ref$tu)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "37cc1f6b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sea_clst_grouped$strand <- seq_bed$strand[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "sea_clst_grouped$chr <- seq_bed$chr[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "sea_clst_grouped$range_start <- seq_bed$start[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "sea_clst_grouped$range_end <- seq_bed$end[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "\n",
    "\n",
    "sea_clst_grouped$most_upstream_motif_start <- apply(sea_clst_grouped,1,function(x){\n",
    "  strand <- x['strand']\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  min_start <- as.numeric(x['min_start'])\n",
    "  chr <- x['chr']\n",
    "  start <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==min_start),'start'] \n",
    "  return(start)\n",
    "})\n",
    "\n",
    "sea_clst_grouped$most_upstream_motif <- apply(sea_clst_grouped,1,function(x){\n",
    "  strand <- x['strand']\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  min_start <- as.numeric(x['min_start'])\n",
    "  chr <- x['chr']\n",
    "  motif <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==min_start),'meme_name']\n",
    "  return(motif)\n",
    "})\n",
    "\n",
    "sea_clst_grouped$most_downstream_motif_start <- apply(sea_clst_grouped,1,function(x){\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  max_end <- as.numeric(x['max_end'])\n",
    "  chr <- x['chr']\n",
    "  start <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==max_end),'start']\n",
    "  return(start)\n",
    "})\n",
    "\n",
    "sea_clst_grouped$most_downstream_motif <- apply(sea_clst_grouped,1,function(x){\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  max_end <- as.numeric(x['max_end'])\n",
    "  chr <- x['chr']\n",
    "  motif <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==max_end),'meme_name']\n",
    "  return(motif)\n",
    "})\n",
    "\n",
    "\n",
    "sea_clst_grouped <- sea_clst_grouped %>%\n",
    "  relocate(chr, .after = seq_name) %>%\n",
    "  relocate(strand, .after = chr) %>%\n",
    "  relocate(range_start, .after = strand) %>%\n",
    "  relocate(range_end, .after = range_start) %>%\n",
    "  relocate(meme_seqs, .after = range_end) %>%\n",
    "  relocate(full_motif_cluster, .after = meme_seqs) %>%\n",
    "  relocate(most_upstream_motif, .after = full_motif_cluster) %>%\n",
    "  relocate(most_upstream_motif_start, .after = most_upstream_motif) %>%\n",
    "  relocate(most_downstream_motif, .after = most_upstream_motif_start) %>%\n",
    "  relocate(most_downstream_motif_start, .after = most_downstream_motif) %>%\n",
    "  relocate(cluster_count, .after = most_downstream_motif_start) %>%\n",
    "  relocate(tu, .before = seq_name) %>% \n",
    "  relocate(gene, .before = tu) %>% select(-c(combinations,cluster_count,overlap_group,min_start,max_end))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "42143c57",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene</th><th scope=col>tu</th><th scope=col>seq_name</th><th scope=col>chr</th><th scope=col>strand</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>meme_seqs</th><th scope=col>full_motif_cluster</th><th scope=col>most_upstream_motif</th><th scope=col>most_upstream_motif_start</th><th scope=col>most_downstream_motif</th><th scope=col>most_downstream_motif_start</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR1</td><td>chr12</td><td>+</td><td>  8875461</td><td>  8875815</td><td>MEME-5,MEME-1,MEME-3              </td><td>FALSE</td><td>MEME-5</td><td>  8875487</td><td>MEME-3</td><td>  8875701</td></tr>\n",
       "\t<tr><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR2</td><td>chr12</td><td>+</td><td>  8876058</td><td>  8876788</td><td>MEME-3,MEME-1,MEME-5              </td><td>FALSE</td><td>MEME-3</td><td>  8876388</td><td>MEME-5</td><td>  8876606</td></tr>\n",
       "\t<tr><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR1 </td><td>chr2 </td><td>-</td><td> 69481056</td><td> 69482589</td><td>MEME-5,MEME-2,MEME-1,MEME-3       </td><td>FALSE</td><td>MEME-5</td><td> 69482286</td><td>MEME-3</td><td> 69482067</td></tr>\n",
       "\t<tr><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR2 </td><td>chr2 </td><td>-</td><td> 69457997</td><td> 69475872</td><td>MEME-3,MEME-1,MEME-2,MEME-5,MEME-4</td><td> TRUE</td><td>MEME-3</td><td> 69459527</td><td>MEME-4</td><td> 69459270</td></tr>\n",
       "\t<tr><td>AAMDC</td><td>TU20683</td><td>TU20683_UTR0</td><td>chr11</td><td>+</td><td> 77918097</td><td> 77918433</td><td>MEME-3,MEME-2,MEME-5,MEME-4       </td><td>FALSE</td><td>MEME-3</td><td> 77918146</td><td>MEME-4</td><td> 77918387</td></tr>\n",
       "\t<tr><td>AARD </td><td>TU16086</td><td>TU16086_UTR0</td><td>chr8 </td><td>+</td><td>116942699</td><td>116944488</td><td>MEME-3,MEME-1,MEME-2,MEME-5,MEME-4</td><td> TRUE</td><td>MEME-3</td><td>116942715</td><td>MEME-4</td><td>116942970</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " gene & tu & seq\\_name & chr & strand & range\\_start & range\\_end & meme\\_seqs & full\\_motif\\_cluster & most\\_upstream\\_motif & most\\_upstream\\_motif\\_start & most\\_downstream\\_motif & most\\_downstream\\_motif\\_start\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <lgl> & <chr> & <int> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t A2ML1 & TU22196 & TU22196\\_UTR1 & chr12 & + &   8875461 &   8875815 & MEME-5,MEME-1,MEME-3               & FALSE & MEME-5 &   8875487 & MEME-3 &   8875701\\\\\n",
       "\t A2ML1 & TU22196 & TU22196\\_UTR2 & chr12 & + &   8876058 &   8876788 & MEME-3,MEME-1,MEME-5               & FALSE & MEME-3 &   8876388 & MEME-5 &   8876606\\\\\n",
       "\t AAK1  & TU5339  & TU5339\\_UTR1  & chr2  & - &  69481056 &  69482589 & MEME-5,MEME-2,MEME-1,MEME-3        & FALSE & MEME-5 &  69482286 & MEME-3 &  69482067\\\\\n",
       "\t AAK1  & TU5339  & TU5339\\_UTR2  & chr2  & - &  69457997 &  69475872 & MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 &  TRUE & MEME-3 &  69459527 & MEME-4 &  69459270\\\\\n",
       "\t AAMDC & TU20683 & TU20683\\_UTR0 & chr11 & + &  77918097 &  77918433 & MEME-3,MEME-2,MEME-5,MEME-4        & FALSE & MEME-3 &  77918146 & MEME-4 &  77918387\\\\\n",
       "\t AARD  & TU16086 & TU16086\\_UTR0 & chr8  & + & 116942699 & 116944488 & MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 &  TRUE & MEME-3 & 116942715 & MEME-4 & 116942970\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 13\n",
       "\n",
       "| gene &lt;chr&gt; | tu &lt;chr&gt; | seq_name &lt;chr&gt; | chr &lt;chr&gt; | strand &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | meme_seqs &lt;chr&gt; | full_motif_cluster &lt;lgl&gt; | most_upstream_motif &lt;chr&gt; | most_upstream_motif_start &lt;int&gt; | most_downstream_motif &lt;chr&gt; | most_downstream_motif_start &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A2ML1 | TU22196 | TU22196_UTR1 | chr12 | + |   8875461 |   8875815 | MEME-5,MEME-1,MEME-3               | FALSE | MEME-5 |   8875487 | MEME-3 |   8875701 |\n",
       "| A2ML1 | TU22196 | TU22196_UTR2 | chr12 | + |   8876058 |   8876788 | MEME-3,MEME-1,MEME-5               | FALSE | MEME-3 |   8876388 | MEME-5 |   8876606 |\n",
       "| AAK1  | TU5339  | TU5339_UTR1  | chr2  | - |  69481056 |  69482589 | MEME-5,MEME-2,MEME-1,MEME-3        | FALSE | MEME-5 |  69482286 | MEME-3 |  69482067 |\n",
       "| AAK1  | TU5339  | TU5339_UTR2  | chr2  | - |  69457997 |  69475872 | MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 |  TRUE | MEME-3 |  69459527 | MEME-4 |  69459270 |\n",
       "| AAMDC | TU20683 | TU20683_UTR0 | chr11 | + |  77918097 |  77918433 | MEME-3,MEME-2,MEME-5,MEME-4        | FALSE | MEME-3 |  77918146 | MEME-4 |  77918387 |\n",
       "| AARD  | TU16086 | TU16086_UTR0 | chr8  | + | 116942699 | 116944488 | MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 |  TRUE | MEME-3 | 116942715 | MEME-4 | 116942970 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene  tu      seq_name     chr   strand range_start range_end\n",
       "1 A2ML1 TU22196 TU22196_UTR1 chr12 +        8875461     8875815\n",
       "2 A2ML1 TU22196 TU22196_UTR2 chr12 +        8876058     8876788\n",
       "3 AAK1  TU5339  TU5339_UTR1  chr2  -       69481056    69482589\n",
       "4 AAK1  TU5339  TU5339_UTR2  chr2  -       69457997    69475872\n",
       "5 AAMDC TU20683 TU20683_UTR0 chr11 +       77918097    77918433\n",
       "6 AARD  TU16086 TU16086_UTR0 chr8  +      116942699   116944488\n",
       "  meme_seqs                          full_motif_cluster most_upstream_motif\n",
       "1 MEME-5,MEME-1,MEME-3               FALSE              MEME-5             \n",
       "2 MEME-3,MEME-1,MEME-5               FALSE              MEME-3             \n",
       "3 MEME-5,MEME-2,MEME-1,MEME-3        FALSE              MEME-5             \n",
       "4 MEME-3,MEME-1,MEME-2,MEME-5,MEME-4  TRUE              MEME-3             \n",
       "5 MEME-3,MEME-2,MEME-5,MEME-4        FALSE              MEME-3             \n",
       "6 MEME-3,MEME-1,MEME-2,MEME-5,MEME-4  TRUE              MEME-3             \n",
       "  most_upstream_motif_start most_downstream_motif most_downstream_motif_start\n",
       "1   8875487                 MEME-3                  8875701                  \n",
       "2   8876388                 MEME-5                  8876606                  \n",
       "3  69482286                 MEME-3                 69482067                  \n",
       "4  69459527                 MEME-4                 69459270                  \n",
       "5  77918146                 MEME-4                 77918387                  \n",
       "6 116942715                 MEME-4                116942970                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sea_clst_grouped <- sea_clst_grouped %>%\n",
    "  # Extract TU and current UTR index for sorting\n",
    "  mutate(\n",
    "    tu_part = gsub('_UTR[0-9]+', '', seq_name),\n",
    "    current_utr = as.numeric(gsub('.*_UTR([0-9]+)', '\\\\1', seq_name))\n",
    "  ) %>%\n",
    "  group_by(tu_part) %>%\n",
    "  arrange(\n",
    "    tu_part, \n",
    "    case_when(\n",
    "      strand == '+' ~ range_start,\n",
    "      strand == '-' ~ -range_end\n",
    "    )\n",
    "  ) %>%\n",
    "  mutate(\n",
    "    # Create unique UTR identifier based on coordinates\n",
    "    # utr_coords = paste(range_start, range_end, sep = \"_\"),\n",
    "    # Get unique coordinates in order\n",
    "    unique_seqs = list(unique(seq_name)),\n",
    "    # Assign index based on position of coordinates in unique list\n",
    "    new_utr_index = case_when(\n",
    "      length(seq_name) == 1 ~ 0,\n",
    "      TRUE ~ match(seq_name, unique_seqs[[1]]) \n",
    "    ),\n",
    "    # Create new seq_name with reindexed UTR\n",
    "    seq_name = paste0(tu_part, '_UTR', new_utr_index)\n",
    "  ) %>%\n",
    "  ungroup() %>%\n",
    "  select(-tu_part, -current_utr, -new_utr_index, -unique_seqs) %>%\n",
    "  arrange(gene, tu, seq_name)\n",
    "\n",
    "head(sea_clst_grouped)\n",
    "write.table(sea_clst_grouped,\n",
    "            file = paste0(filedir,'/','motif_cluster_detection_res_',fprefix,'.tsv'),\n",
    "            sep = '\\t',quote = F,row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b2d999ad",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 16</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seq_name</th><th scope=col>site_strand</th><th scope=col>site_start</th><th scope=col>pval</th><th scope=col>flank1</th><th scope=col>site</th><th scope=col>flank2</th><th scope=col>meme_name</th><th scope=col>strand</th><th scope=col>cis</th><th scope=col>chr</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>site_width</th><th scope=col>start</th><th scope=col>end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1359</th><td>TU17139_UTR8420</td><td>-</td><td>1889</td><td>NA</td><td>NA</td><td>TGCAATGGCACGATCTCGGCTCACTGCAGCCT               </td><td>NA</td><td>MEME-5</td><td>+</td><td>FALSE</td><td>chr9</td><td>32450109</td><td>32454770</td><td>32</td><td>32451997</td><td>32452029</td></tr>\n",
       "\t<tr><th scope=row>24591</th><td>TU17139_UTR8420</td><td>+</td><td>1669</td><td>NA</td><td>NA</td><td>AACAGCCAGGCGCAGTGGCTCATGCCTATAATTCCATCAC       </td><td>NA</td><td>MEME-3</td><td>+</td><td> TRUE</td><td>chr9</td><td>32450109</td><td>32454770</td><td>40</td><td>32451777</td><td>32451817</td></tr>\n",
       "\t<tr><th scope=row>28865</th><td>TU17139_UTR8420</td><td>+</td><td>1748</td><td>NA</td><td>NA</td><td>GTTCTAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAA</td><td>NA</td><td>MEME-1</td><td>+</td><td> TRUE</td><td>chr9</td><td>32450109</td><td>32454770</td><td>47</td><td>32451856</td><td>32451903</td></tr>\n",
       "\t<tr><th scope=row>35297</th><td>TU17139_UTR8420</td><td>+</td><td>2816</td><td>NA</td><td>NA</td><td>GCCTGGGCAATAGAGTGAGACCCTATCAATCAATCACCACC      </td><td>NA</td><td>MEME-4</td><td>+</td><td> TRUE</td><td>chr9</td><td>32450109</td><td>32454770</td><td>41</td><td>32452924</td><td>32452965</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 16\n",
       "\\begin{tabular}{r|llllllllllllllll}\n",
       "  & seq\\_name & site\\_strand & site\\_start & pval & flank1 & site & flank2 & meme\\_name & strand & cis & chr & range\\_start & range\\_end & site\\_width & start & end\\\\\n",
       "  & <chr> & <chr> & <int> & <lgl> & <lgl> & <chr> & <lgl> & <chr> & <chr> & <lgl> & <chr> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1359 & TU17139\\_UTR8420 & - & 1889 & NA & NA & TGCAATGGCACGATCTCGGCTCACTGCAGCCT                & NA & MEME-5 & + & FALSE & chr9 & 32450109 & 32454770 & 32 & 32451997 & 32452029\\\\\n",
       "\t24591 & TU17139\\_UTR8420 & + & 1669 & NA & NA & AACAGCCAGGCGCAGTGGCTCATGCCTATAATTCCATCAC        & NA & MEME-3 & + &  TRUE & chr9 & 32450109 & 32454770 & 40 & 32451777 & 32451817\\\\\n",
       "\t28865 & TU17139\\_UTR8420 & + & 1748 & NA & NA & GTTCTAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAA & NA & MEME-1 & + &  TRUE & chr9 & 32450109 & 32454770 & 47 & 32451856 & 32451903\\\\\n",
       "\t35297 & TU17139\\_UTR8420 & + & 2816 & NA & NA & GCCTGGGCAATAGAGTGAGACCCTATCAATCAATCACCACC       & NA & MEME-4 & + &  TRUE & chr9 & 32450109 & 32454770 & 41 & 32452924 & 32452965\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 16\n",
       "\n",
       "| <!--/--> | seq_name &lt;chr&gt; | site_strand &lt;chr&gt; | site_start &lt;int&gt; | pval &lt;lgl&gt; | flank1 &lt;lgl&gt; | site &lt;chr&gt; | flank2 &lt;lgl&gt; | meme_name &lt;chr&gt; | strand &lt;chr&gt; | cis &lt;lgl&gt; | chr &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | site_width &lt;int&gt; | start &lt;int&gt; | end &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1359 | TU17139_UTR8420 | - | 1889 | NA | NA | TGCAATGGCACGATCTCGGCTCACTGCAGCCT                | NA | MEME-5 | + | FALSE | chr9 | 32450109 | 32454770 | 32 | 32451997 | 32452029 |\n",
       "| 24591 | TU17139_UTR8420 | + | 1669 | NA | NA | AACAGCCAGGCGCAGTGGCTCATGCCTATAATTCCATCAC        | NA | MEME-3 | + |  TRUE | chr9 | 32450109 | 32454770 | 40 | 32451777 | 32451817 |\n",
       "| 28865 | TU17139_UTR8420 | + | 1748 | NA | NA | GTTCTAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAA | NA | MEME-1 | + |  TRUE | chr9 | 32450109 | 32454770 | 47 | 32451856 | 32451903 |\n",
       "| 35297 | TU17139_UTR8420 | + | 2816 | NA | NA | GCCTGGGCAATAGAGTGAGACCCTATCAATCAATCACCACC       | NA | MEME-4 | + |  TRUE | chr9 | 32450109 | 32454770 | 41 | 32452924 | 32452965 |\n",
       "\n"
      ],
      "text/plain": [
       "      seq_name        site_strand site_start pval flank1\n",
       "1359  TU17139_UTR8420 -           1889       NA   NA    \n",
       "24591 TU17139_UTR8420 +           1669       NA   NA    \n",
       "28865 TU17139_UTR8420 +           1748       NA   NA    \n",
       "35297 TU17139_UTR8420 +           2816       NA   NA    \n",
       "      site                                            flank2 meme_name strand\n",
       "1359  TGCAATGGCACGATCTCGGCTCACTGCAGCCT                NA     MEME-5    +     \n",
       "24591 AACAGCCAGGCGCAGTGGCTCATGCCTATAATTCCATCAC        NA     MEME-3    +     \n",
       "28865 GTTCTAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAA NA     MEME-1    +     \n",
       "35297 GCCTGGGCAATAGAGTGAGACCCTATCAATCAATCACCACC       NA     MEME-4    +     \n",
       "      cis   chr  range_start range_end site_width start    end     \n",
       "1359  FALSE chr9 32450109    32454770  32         32451997 32452029\n",
       "24591  TRUE chr9 32450109    32454770  40         32451777 32451817\n",
       "28865  TRUE chr9 32450109    32454770  47         32451856 32451903\n",
       "35297  TRUE chr9 32450109    32454770  41         32452924 32452965"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sea_res[grepl('^TU17139',sea_res$seq_name),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7fe9ddc6",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene</th><th scope=col>tu</th><th scope=col>seq_name</th><th scope=col>chr</th><th scope=col>strand</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>meme_seqs</th><th scope=col>full_motif_cluster</th><th scope=col>most_upstream_motif</th><th scope=col>most_upstream_motif_start</th><th scope=col>most_downstream_motif</th><th scope=col>most_downstream_motif_start</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ACO1</td><td>TU17139</td><td>TU17139_UTR0</td><td>chr9</td><td>+</td><td>32450109</td><td>32454770</td><td>MEME-3,MEME-1,MEME-5</td><td>FALSE</td><td>MEME-3</td><td>32451777</td><td>MEME-5</td><td>32451997</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " gene & tu & seq\\_name & chr & strand & range\\_start & range\\_end & meme\\_seqs & full\\_motif\\_cluster & most\\_upstream\\_motif & most\\_upstream\\_motif\\_start & most\\_downstream\\_motif & most\\_downstream\\_motif\\_start\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <lgl> & <chr> & <int> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t ACO1 & TU17139 & TU17139\\_UTR0 & chr9 & + & 32450109 & 32454770 & MEME-3,MEME-1,MEME-5 & FALSE & MEME-3 & 32451777 & MEME-5 & 32451997\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 × 13\n",
       "\n",
       "| gene &lt;chr&gt; | tu &lt;chr&gt; | seq_name &lt;chr&gt; | chr &lt;chr&gt; | strand &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | meme_seqs &lt;chr&gt; | full_motif_cluster &lt;lgl&gt; | most_upstream_motif &lt;chr&gt; | most_upstream_motif_start &lt;int&gt; | most_downstream_motif &lt;chr&gt; | most_downstream_motif_start &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ACO1 | TU17139 | TU17139_UTR0 | chr9 | + | 32450109 | 32454770 | MEME-3,MEME-1,MEME-5 | FALSE | MEME-3 | 32451777 | MEME-5 | 32451997 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene tu      seq_name     chr  strand range_start range_end\n",
       "1 ACO1 TU17139 TU17139_UTR0 chr9 +      32450109    32454770 \n",
       "  meme_seqs            full_motif_cluster most_upstream_motif\n",
       "1 MEME-3,MEME-1,MEME-5 FALSE              MEME-3             \n",
       "  most_upstream_motif_start most_downstream_motif most_downstream_motif_start\n",
       "1 32451777                  MEME-5                32451997                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# sea_clst_grouped[sea_clst_grouped$strand=='-',]%>% group_by(tu) %>% summarize(n = n_distinct(seq_name)) %>% arrange(desc(n)) \n",
    "sea_clst_grouped[sea_clst_grouped$tu == 'TU17139',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4a30a5e5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "3791"
      ],
      "text/latex": [
       "3791"
      ],
      "text/markdown": [
       "3791"
      ],
      "text/plain": [
       "[1] 3791"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Number of TUs with lengthening motif clusters\n",
    "length(unique(sea_clst_grouped$tu))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f74e5f76",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "4822"
      ],
      "text/latex": [
       "4822"
      ],
      "text/markdown": [
       "4822"
      ],
      "text/plain": [
       "[1] 4822"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sea_clst_grouped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8b6a634f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>tu</th><th scope=col>seq_name</th><th scope=col>chr</th><th scope=col>strand</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>meme_seqs</th><th scope=col>full_motif_cluster</th><th scope=col>most_upstream_motif</th><th scope=col>most_upstream_motif_start</th><th scope=col>most_downstream_motif</th><th scope=col>most_downstream_motif_start</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR1</td><td>chr12</td><td>+</td><td>  8875461</td><td>  8875815</td><td>MEME-5,MEME-1,MEME-3              </td><td>FALSE</td><td>MEME-5</td><td>  8875487</td><td>MEME-3</td><td>  8875701</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR2</td><td>chr12</td><td>+</td><td>  8876058</td><td>  8876788</td><td>MEME-3,MEME-1,MEME-5              </td><td>FALSE</td><td>MEME-3</td><td>  8876388</td><td>MEME-5</td><td>  8876606</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR1 </td><td>chr2 </td><td>-</td><td> 69481056</td><td> 69482589</td><td>MEME-5,MEME-2,MEME-1,MEME-3       </td><td>FALSE</td><td>MEME-5</td><td> 69482286</td><td>MEME-3</td><td> 69482067</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR2 </td><td>chr2 </td><td>-</td><td> 69457997</td><td> 69475872</td><td>MEME-3,MEME-1,MEME-2,MEME-5,MEME-4</td><td> TRUE</td><td>MEME-3</td><td> 69459527</td><td>MEME-4</td><td> 69459270</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>AAMDC</td><td>TU20683</td><td>TU20683_UTR0</td><td>chr11</td><td>+</td><td> 77918097</td><td> 77918433</td><td>MEME-3,MEME-2,MEME-5,MEME-4       </td><td>FALSE</td><td>MEME-3</td><td> 77918146</td><td>MEME-4</td><td> 77918387</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>AARD </td><td>TU16086</td><td>TU16086_UTR0</td><td>chr8 </td><td>+</td><td>116942699</td><td>116944488</td><td>MEME-3,MEME-1,MEME-2,MEME-5,MEME-4</td><td> TRUE</td><td>MEME-3</td><td>116942715</td><td>MEME-4</td><td>116942970</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 13\n",
       "\\begin{tabular}{r|lllllllllllll}\n",
       "  & gene & tu & seq\\_name & chr & strand & range\\_start & range\\_end & meme\\_seqs & full\\_motif\\_cluster & most\\_upstream\\_motif & most\\_upstream\\_motif\\_start & most\\_downstream\\_motif & most\\_downstream\\_motif\\_start\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <lgl> & <chr> & <int> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & A2ML1 & TU22196 & TU22196\\_UTR1 & chr12 & + &   8875461 &   8875815 & MEME-5,MEME-1,MEME-3               & FALSE & MEME-5 &   8875487 & MEME-3 &   8875701\\\\\n",
       "\t2 & A2ML1 & TU22196 & TU22196\\_UTR2 & chr12 & + &   8876058 &   8876788 & MEME-3,MEME-1,MEME-5               & FALSE & MEME-3 &   8876388 & MEME-5 &   8876606\\\\\n",
       "\t3 & AAK1  & TU5339  & TU5339\\_UTR1  & chr2  & - &  69481056 &  69482589 & MEME-5,MEME-2,MEME-1,MEME-3        & FALSE & MEME-5 &  69482286 & MEME-3 &  69482067\\\\\n",
       "\t4 & AAK1  & TU5339  & TU5339\\_UTR2  & chr2  & - &  69457997 &  69475872 & MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 &  TRUE & MEME-3 &  69459527 & MEME-4 &  69459270\\\\\n",
       "\t5 & AAMDC & TU20683 & TU20683\\_UTR0 & chr11 & + &  77918097 &  77918433 & MEME-3,MEME-2,MEME-5,MEME-4        & FALSE & MEME-3 &  77918146 & MEME-4 &  77918387\\\\\n",
       "\t6 & AARD  & TU16086 & TU16086\\_UTR0 & chr8  & + & 116942699 & 116944488 & MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 &  TRUE & MEME-3 & 116942715 & MEME-4 & 116942970\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 13\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | tu &lt;chr&gt; | seq_name &lt;chr&gt; | chr &lt;chr&gt; | strand &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | meme_seqs &lt;chr&gt; | full_motif_cluster &lt;lgl&gt; | most_upstream_motif &lt;chr&gt; | most_upstream_motif_start &lt;int&gt; | most_downstream_motif &lt;chr&gt; | most_downstream_motif_start &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A2ML1 | TU22196 | TU22196_UTR1 | chr12 | + |   8875461 |   8875815 | MEME-5,MEME-1,MEME-3               | FALSE | MEME-5 |   8875487 | MEME-3 |   8875701 |\n",
       "| 2 | A2ML1 | TU22196 | TU22196_UTR2 | chr12 | + |   8876058 |   8876788 | MEME-3,MEME-1,MEME-5               | FALSE | MEME-3 |   8876388 | MEME-5 |   8876606 |\n",
       "| 3 | AAK1  | TU5339  | TU5339_UTR1  | chr2  | - |  69481056 |  69482589 | MEME-5,MEME-2,MEME-1,MEME-3        | FALSE | MEME-5 |  69482286 | MEME-3 |  69482067 |\n",
       "| 4 | AAK1  | TU5339  | TU5339_UTR2  | chr2  | - |  69457997 |  69475872 | MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 |  TRUE | MEME-3 |  69459527 | MEME-4 |  69459270 |\n",
       "| 5 | AAMDC | TU20683 | TU20683_UTR0 | chr11 | + |  77918097 |  77918433 | MEME-3,MEME-2,MEME-5,MEME-4        | FALSE | MEME-3 |  77918146 | MEME-4 |  77918387 |\n",
       "| 6 | AARD  | TU16086 | TU16086_UTR0 | chr8  | + | 116942699 | 116944488 | MEME-3,MEME-1,MEME-2,MEME-5,MEME-4 |  TRUE | MEME-3 | 116942715 | MEME-4 | 116942970 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene  tu      seq_name     chr   strand range_start range_end\n",
       "1 A2ML1 TU22196 TU22196_UTR1 chr12 +        8875461     8875815\n",
       "2 A2ML1 TU22196 TU22196_UTR2 chr12 +        8876058     8876788\n",
       "3 AAK1  TU5339  TU5339_UTR1  chr2  -       69481056    69482589\n",
       "4 AAK1  TU5339  TU5339_UTR2  chr2  -       69457997    69475872\n",
       "5 AAMDC TU20683 TU20683_UTR0 chr11 +       77918097    77918433\n",
       "6 AARD  TU16086 TU16086_UTR0 chr8  +      116942699   116944488\n",
       "  meme_seqs                          full_motif_cluster most_upstream_motif\n",
       "1 MEME-5,MEME-1,MEME-3               FALSE              MEME-5             \n",
       "2 MEME-3,MEME-1,MEME-5               FALSE              MEME-3             \n",
       "3 MEME-5,MEME-2,MEME-1,MEME-3        FALSE              MEME-5             \n",
       "4 MEME-3,MEME-1,MEME-2,MEME-5,MEME-4  TRUE              MEME-3             \n",
       "5 MEME-3,MEME-2,MEME-5,MEME-4        FALSE              MEME-3             \n",
       "6 MEME-3,MEME-1,MEME-2,MEME-5,MEME-4  TRUE              MEME-3             \n",
       "  most_upstream_motif_start most_downstream_motif most_downstream_motif_start\n",
       "1   8875487                 MEME-3                  8875701                  \n",
       "2   8876388                 MEME-5                  8876606                  \n",
       "3  69482286                 MEME-3                 69482067                  \n",
       "4  69459527                 MEME-4                 69459270                  \n",
       "5  77918146                 MEME-4                 77918387                  \n",
       "6 116942715                 MEME-4                116942970                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of full or partial lengthening motif cluster: 4822 \n",
      "Number of full or partial lengthening motif cluster outside discovery set: 4797 \n",
      "Number of TUs with full or partial lengthening motif cluster: 3791 \n",
      "Number of TUs with full or partial lengthening motif cluster outside discovery set: 3772 \n"
     ]
    }
   ],
   "source": [
    "filedir=paste0('sea_out_lgthn_',fprefix)\n",
    "sea_clst_grouped <- read.delim(paste0(filedir,'/','motif_cluster_detection_res_', fprefix, '.tsv'))\n",
    "sea_clst_grouped_outside_discovery_set <- sea_clst_grouped[!(sea_clst_grouped$tu %in% turef_lgthn_mc_genes$tu),]\n",
    "write.table(sea_clst_grouped_outside_discovery_set,\n",
    "            file = paste0(filedir,'/','motif_cluster_detection_res_outside_discovery_set_',fprefix,'.tsv'),\n",
    "            sep = '\\t',quote = F,row.names = F)\n",
    "head(sea_clst_grouped_outside_discovery_set)\n",
    "cat('Number of full or partial lengthening motif cluster:', nrow(sea_clst_grouped), '\\n')\n",
    "cat('Number of full or partial lengthening motif cluster outside discovery set:', nrow(sea_clst_grouped_outside_discovery_set), '\\n')\n",
    "cat('Number of TUs with full or partial lengthening motif cluster:', length(unique(sea_clst_grouped$tu)), '\\n')\n",
    "cat('Number of TUs with full or partial lengthening motif cluster outside discovery set:', length(unique(sea_clst_grouped_outside_discovery_set$tu)), '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4795f9c3",
   "metadata": {},
   "source": [
    "##### Shortening motif cluster detecting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ccd6c205",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running sea cluster detection for sea_out_shrtn_all_TU_disjoint_utr3...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 20 × 2\u001b[39m\n",
      "   seq_name             n\n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m            \u001b[3m\u001b[90m<int>\u001b[39m\u001b[23m\n",
      "\u001b[90m 1\u001b[39m TU22647_UTR12934    97\n",
      "\u001b[90m 2\u001b[39m TU4659_UTR23842     76\n",
      "\u001b[90m 3\u001b[39m TU26609_UTR15674    73\n",
      "\u001b[90m 4\u001b[39m TU26145_UTR15390    68\n",
      "\u001b[90m 5\u001b[39m TU32685_UTR20212    62\n",
      "\u001b[90m 6\u001b[39m TU36069_UTR25042    59\n",
      "\u001b[90m 7\u001b[39m TU13708_UTR5877     56\n",
      "\u001b[90m 8\u001b[39m TU14230_UTR6295     51\n",
      "\u001b[90m 9\u001b[39m TU27563_UTR16332    51\n",
      "\u001b[90m10\u001b[39m TU3786_UTR23194     51\n",
      "\u001b[90m11\u001b[39m TU12396_UTR4853     50\n",
      "\u001b[90m12\u001b[39m TU13562_UTR5805     50\n",
      "\u001b[90m13\u001b[39m TU23086_UTR13233    50\n",
      "\u001b[90m14\u001b[39m TU418_UTR348        50\n",
      "\u001b[90m15\u001b[39m TU9714_UTR29577     49\n",
      "\u001b[90m16\u001b[39m TU22761_UTR12995    47\n",
      "\u001b[90m17\u001b[39m TU37991_UTR26666    47\n",
      "\u001b[90m18\u001b[39m TU1054_UTR867       46\n",
      "\u001b[90m19\u001b[39m TU33335_UTR20846    46\n",
      "\u001b[90m20\u001b[39m TU28128_UTR16802    45\n",
      "Removing problematic seq TU22647_UTR12934 \n",
      "  |++++++++++++++++++++++++++++++++++++++++++++++++++| 100%\n"
     ]
    }
   ],
   "source": [
    "#### SEA\n",
    "filedir=paste0('sea_out_shrtn_',fprefix)\n",
    "message(paste0('Running sea cluster detection for ',filedir,'...'))\n",
    "  \n",
    "sea_res <- read.delim(paste0(filedir,'/sites.tsv'))\n",
    "sea_res <- sea_res[!sea_res$motif_ID=='#',]\n",
    "sea_res$strand <- str_extract_all(sea_res$seq_ID, \"\\\\(([^)]+)\\\\)\") %>% unlist\n",
    "sea_res$strand <- gsub(\"\\\\(|\\\\)\",\"\",sea_res$strand)\n",
    "sea_res$cis <- sea_res$site_Strand==sea_res$strand\n",
    "sea_res$seq_ID <- gsub(\"\\\\(([^)]+)\\\\)\",\"\",sea_res$seq_ID)\n",
    "sea_res$chr <- seq_bed$chr[match(sea_res$seq_ID,seq_bed$seq_name)]\n",
    "sea_res$range_start <- seq_bed$start[match(sea_res$seq_ID,seq_bed$seq_name)]\n",
    "sea_res$range_end <- seq_bed$end[match(sea_res$seq_ID,seq_bed$seq_name)]\n",
    "sea_res$site_width <- nchar(sea_res$site_Sequence)\n",
    "sea_res$start <- ifelse(sea_res$strand=='+',sea_res$range_start+sea_res$site_Start-1,\n",
    "                                                sea_res$range_end-sea_res$site_Start-sea_res$site_width+1)\n",
    "sea_res$end <- sea_res$start+sea_res$site_width\n",
    "colnames(sea_res)[c(2,3,4,5,6,8)] <- c('meme_name','seq_name','site_start','site_end','site_strand','site')\n",
    "sea_res <- sea_res %>% select(seq_name,site_strand,site_start,site,meme_name,strand,cis,chr,\n",
    "                                                      range_start,range_end,site_width,start,end)\n",
    "sea_res <- sea_res %>% mutate(pval = NA,flank1 = NA,flank2 = NA) %>%\n",
    "    relocate(pval, .after = site_start) %>%\n",
    "    relocate(flank1, .after = pval) %>%\n",
    "    relocate(flank2, .after = site)\n",
    "top_nmotif <- sea_res %>% group_by(seq_name) %>% summarize(n=n()) %>% arrange(desc(n))\n",
    "print(top_nmotif[1:20,])\n",
    "write.table(sea_res,paste0(filedir,'/sea_res_processed_',fprefix,'.tsv'),\n",
    "            sep = '\\t',quote = F,row.names = F)   \n",
    "\n",
    "# remove problematic seq\n",
    "seq_names <- unique(sea_res$seq_name)\n",
    "problematic_seq_idx <- 1518\n",
    "# problematic_seq_idx <- which(seq_names %in% top_nmotif$seq_name[1:20])\n",
    "cat('Removing problematic seq', seq_names[problematic_seq_idx], '\\n')\n",
    "subset_seq_names <- seq_names[-(problematic_seq_idx)]\n",
    "sea_res <- sea_res[sea_res$seq_name %in% subset_seq_names,]\n",
    "\n",
    "sea_clst_found <- find_motif_cluster_in_sea(sea_res,mc = c('MEME-1'='-','MEME-2'='-','MEME-4'='-','MEME-3'='+','MEME-5'='-'),\n",
    "                                              min_nmotif=3,btwn_mt_max_gap=40,ncores = 1)\n",
    "saveRDS(sea_clst_found,paste0(filedir,'/sea_clst_found_',fprefix,'.rds'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "5fd7302c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              seq_name site_strand site_start meme_name strand   cis   chr\n",
      "8364  TU22647_UTR12934           -        108    MEME-1      + FALSE chr12\n",
      "16113 TU22647_UTR12934           -        155    MEME-2      + FALSE chr12\n",
      "34141 TU22647_UTR12934           -        224    MEME-4      + FALSE chr12\n",
      "5200  TU22647_UTR12934           -        242    MEME-1      + FALSE chr12\n",
      "22095 TU22647_UTR12934           +        275    MEME-3      +  TRUE chr12\n",
      "31960 TU22647_UTR12934           -        341    MEME-5      + FALSE chr12\n",
      "12335 TU22647_UTR12934           -        573    MEME-1      + FALSE chr12\n",
      "25474 TU22647_UTR12934           +        606    MEME-3      +  TRUE chr12\n",
      "6964  TU22647_UTR12934           -       1338    MEME-1      + FALSE chr12\n",
      "33894 TU22647_UTR12934           -       1454    MEME-4      + FALSE chr12\n",
      "6734  TU22647_UTR12934           -       1472    MEME-1      + FALSE chr12\n",
      "22742 TU22647_UTR12934           +       1505    MEME-3      +  TRUE chr12\n",
      "29782 TU22647_UTR12934           -       1571    MEME-5      + FALSE chr12\n",
      "27545 TU22647_UTR12934           -       1765    MEME-3      + FALSE chr12\n",
      "12776 TU22647_UTR12934           +       1800    MEME-1      +  TRUE chr12\n",
      "20730 TU22647_UTR12934           +       1888    MEME-2      +  TRUE chr12\n",
      "9684  TU22647_UTR12934           +       1935    MEME-1      +  TRUE chr12\n",
      "11846 TU22647_UTR12934           -       2503    MEME-1      + FALSE chr12\n",
      "26654 TU22647_UTR12934           +       2536    MEME-3      +  TRUE chr12\n",
      "2621  TU22647_UTR12934           -       3140    MEME-1      + FALSE chr12\n",
      "19155 TU22647_UTR12934           -       3187    MEME-2      + FALSE chr12\n",
      "6342  TU22647_UTR12934           -       3275    MEME-1      + FALSE chr12\n",
      "25078 TU22647_UTR12934           +       3308    MEME-3      +  TRUE chr12\n",
      "33463 TU22647_UTR12934           -       3374    MEME-5      + FALSE chr12\n",
      "3919  TU22647_UTR12934           -       4093    MEME-1      + FALSE chr12\n",
      "16298 TU22647_UTR12934           -       4140    MEME-2      + FALSE chr12\n",
      "34951 TU22647_UTR12934           -       4210    MEME-4      + FALSE chr12\n",
      "8697  TU22647_UTR12934           -       4228    MEME-1      + FALSE chr12\n",
      "21747 TU22647_UTR12934           +       4261    MEME-3      +  TRUE chr12\n",
      "32073 TU22647_UTR12934           -       4326    MEME-5      + FALSE chr12\n",
      "3080  TU22647_UTR12934           -       4508    MEME-1      + FALSE chr12\n",
      "14632 TU22647_UTR12934           -       4555    MEME-2      + FALSE chr12\n",
      "6613  TU22647_UTR12934           -       4645    MEME-1      + FALSE chr12\n",
      "26626 TU22647_UTR12934           +       4678    MEME-3      +  TRUE chr12\n",
      "2924  TU22647_UTR12934           -       5685    MEME-1      + FALSE chr12\n",
      "9894  TU22647_UTR12934           -       5820    MEME-1      + FALSE chr12\n",
      "33571 TU22647_UTR12934           -       5918    MEME-5      + FALSE chr12\n",
      "1782  TU22647_UTR12934           -       6125    MEME-1      + FALSE chr12\n",
      "14994 TU22647_UTR12934           -       6172    MEME-2      + FALSE chr12\n",
      "6822  TU22647_UTR12934           -       6260    MEME-1      + FALSE chr12\n",
      "25468 TU22647_UTR12934           +       6293    MEME-3      +  TRUE chr12\n",
      "31653 TU22647_UTR12934           -       6358    MEME-5      + FALSE chr12\n",
      "19837 TU22647_UTR12934           -       6718    MEME-2      + FALSE chr12\n",
      "21085 TU22647_UTR12934           -       7218    MEME-3      + FALSE chr12\n",
      "10337 TU22647_UTR12934           +       7253    MEME-1      +  TRUE chr12\n",
      "17907 TU22647_UTR12934           +       7341    MEME-2      +  TRUE chr12\n",
      "6169  TU22647_UTR12934           +       7388    MEME-1      +  TRUE chr12\n",
      "25885 TU22647_UTR12934           -       8587    MEME-3      + FALSE chr12\n",
      "8818  TU22647_UTR12934           +       8622    MEME-1      +  TRUE chr12\n",
      "33013 TU22647_UTR12934           +       8687    MEME-5      +  TRUE chr12\n",
      "4152  TU22647_UTR12934           -       9288    MEME-1      + FALSE chr12\n",
      "19158 TU22647_UTR12934           -       9335    MEME-2      + FALSE chr12\n",
      "7212  TU22647_UTR12934           -       9431    MEME-1      + FALSE chr12\n",
      "16114 TU22647_UTR12934           -       9478    MEME-2      + FALSE chr12\n",
      "8452  TU22647_UTR12934           -       9565    MEME-1      + FALSE chr12\n",
      "22481 TU22647_UTR12934           +       9598    MEME-3      +  TRUE chr12\n",
      "29783 TU22647_UTR12934           -       9664    MEME-5      + FALSE chr12\n",
      "32596 TU22647_UTR12934           -       9822    MEME-5      + FALSE chr12\n",
      "6720  TU22647_UTR12934           -       9997    MEME-1      + FALSE chr12\n",
      "7477  TU22647_UTR12934           +      10196    MEME-1      +  TRUE chr12\n",
      "8014  TU22647_UTR12934           -      10799    MEME-1      + FALSE chr12\n",
      "14633 TU22647_UTR12934           -      10846    MEME-2      + FALSE chr12\n",
      "11541 TU22647_UTR12934           -      10934    MEME-1      + FALSE chr12\n",
      "21776 TU22647_UTR12934           +      10967    MEME-3      +  TRUE chr12\n",
      "30367 TU22647_UTR12934           -      11033    MEME-5      + FALSE chr12\n",
      "2146  TU22647_UTR12934           -      11113    MEME-1      + FALSE chr12\n",
      "16194 TU22647_UTR12934           -      11160    MEME-2      + FALSE chr12\n",
      "31868 TU22647_UTR12934           -      11295    MEME-5      + FALSE chr12\n",
      "2113  TU22647_UTR12934           -      11360    MEME-1      + FALSE chr12\n",
      "16195 TU22647_UTR12934           -      11407    MEME-2      + FALSE chr12\n",
      "27632 TU22647_UTR12934           +      11532    MEME-3      +  TRUE chr12\n",
      "29030 TU22647_UTR12934           -      11597    MEME-5      + FALSE chr12\n",
      "21198 TU22647_UTR12934           +      11705    MEME-3      +  TRUE chr12\n",
      "32981 TU22647_UTR12934           -      11770    MEME-5      + FALSE chr12\n",
      "2006  TU22647_UTR12934           -      12969    MEME-1      + FALSE chr12\n",
      "17431 TU22647_UTR12934           -      13016    MEME-2      + FALSE chr12\n",
      "5928  TU22647_UTR12934           -      13104    MEME-1      + FALSE chr12\n",
      "12321 TU22647_UTR12934           -      13641    MEME-1      + FALSE chr12\n",
      "16870 TU22647_UTR12934           -      13688    MEME-2      + FALSE chr12\n",
      "6546  TU22647_UTR12934           -      13776    MEME-1      + FALSE chr12\n",
      "26747 TU22647_UTR12934           +      13809    MEME-3      +  TRUE chr12\n",
      "30297 TU22647_UTR12934           -      13874    MEME-5      + FALSE chr12\n",
      "2548  TU22647_UTR12934           -      13956    MEME-1      + FALSE chr12\n",
      "17392 TU22647_UTR12934           -      14003    MEME-2      + FALSE chr12\n",
      "9245  TU22647_UTR12934           -      14091    MEME-1      + FALSE chr12\n",
      "25183 TU22647_UTR12934           +      14124    MEME-3      +  TRUE chr12\n",
      "28938 TU22647_UTR12934           -      14190    MEME-5      + FALSE chr12\n",
      "12258 TU22647_UTR12934           +      14334    MEME-1      +  TRUE chr12\n",
      "34309 TU22647_UTR12934           +      14343    MEME-4      +  TRUE chr12\n",
      "33267 TU22647_UTR12934           +      15286    MEME-5      +  TRUE chr12\n",
      "8519  TU22647_UTR12934           +      15384    MEME-1      +  TRUE chr12\n",
      "34447 TU22647_UTR12934           +      15393    MEME-4      +  TRUE chr12\n",
      "4017  TU22647_UTR12934           +      15523    MEME-1      +  TRUE chr12\n",
      "31729 TU22647_UTR12934           +      15604    MEME-5      +  TRUE chr12\n",
      "22466 TU22647_UTR12934           -      15668    MEME-3      + FALSE chr12\n",
      "7341  TU22647_UTR12934           +      15703    MEME-1      +  TRUE chr12\n",
      "34981 TU22647_UTR12934           +      15712    MEME-4      +  TRUE chr12\n",
      "      range_start range_end site_width    start      end\n",
      "8364     68764888  68781469         41 68764995 68765036\n",
      "16113    68764888  68781469         41 68765042 68765083\n",
      "34141    68764888  68781469         50 68765111 68765161\n",
      "5200     68764888  68781469         41 68765129 68765170\n",
      "22095    68764888  68781469         43 68765162 68765205\n",
      "31960    68764888  68781469         41 68765228 68765269\n",
      "12335    68764888  68781469         41 68765460 68765501\n",
      "25474    68764888  68781469         43 68765493 68765536\n",
      "6964     68764888  68781469         41 68766225 68766266\n",
      "33894    68764888  68781469         50 68766341 68766391\n",
      "6734     68764888  68781469         41 68766359 68766400\n",
      "22742    68764888  68781469         43 68766392 68766435\n",
      "29782    68764888  68781469         41 68766458 68766499\n",
      "27545    68764888  68781469         43 68766652 68766695\n",
      "12776    68764888  68781469         41 68766687 68766728\n",
      "20730    68764888  68781469         41 68766775 68766816\n",
      "9684     68764888  68781469         41 68766822 68766863\n",
      "11846    68764888  68781469         41 68767390 68767431\n",
      "26654    68764888  68781469         43 68767423 68767466\n",
      "2621     68764888  68781469         41 68768027 68768068\n",
      "19155    68764888  68781469         41 68768074 68768115\n",
      "6342     68764888  68781469         41 68768162 68768203\n",
      "25078    68764888  68781469         43 68768195 68768238\n",
      "33463    68764888  68781469         41 68768261 68768302\n",
      "3919     68764888  68781469         41 68768980 68769021\n",
      "16298    68764888  68781469         41 68769027 68769068\n",
      "34951    68764888  68781469         50 68769097 68769147\n",
      "8697     68764888  68781469         41 68769115 68769156\n",
      "21747    68764888  68781469         43 68769148 68769191\n",
      "32073    68764888  68781469         41 68769213 68769254\n",
      "3080     68764888  68781469         41 68769395 68769436\n",
      "14632    68764888  68781469         41 68769442 68769483\n",
      "6613     68764888  68781469         41 68769532 68769573\n",
      "26626    68764888  68781469         43 68769565 68769608\n",
      "2924     68764888  68781469         41 68770572 68770613\n",
      "9894     68764888  68781469         41 68770707 68770748\n",
      "33571    68764888  68781469         41 68770805 68770846\n",
      "1782     68764888  68781469         41 68771012 68771053\n",
      "14994    68764888  68781469         41 68771059 68771100\n",
      "6822     68764888  68781469         41 68771147 68771188\n",
      "25468    68764888  68781469         43 68771180 68771223\n",
      "31653    68764888  68781469         41 68771245 68771286\n",
      "19837    68764888  68781469         41 68771605 68771646\n",
      "21085    68764888  68781469         43 68772105 68772148\n",
      "10337    68764888  68781469         41 68772140 68772181\n",
      "17907    68764888  68781469         41 68772228 68772269\n",
      "6169     68764888  68781469         41 68772275 68772316\n",
      "25885    68764888  68781469         43 68773474 68773517\n",
      "8818     68764888  68781469         41 68773509 68773550\n",
      "33013    68764888  68781469         41 68773574 68773615\n",
      "4152     68764888  68781469         41 68774175 68774216\n",
      "19158    68764888  68781469         41 68774222 68774263\n",
      "7212     68764888  68781469         41 68774318 68774359\n",
      "16114    68764888  68781469         41 68774365 68774406\n",
      "8452     68764888  68781469         41 68774452 68774493\n",
      "22481    68764888  68781469         43 68774485 68774528\n",
      "29783    68764888  68781469         41 68774551 68774592\n",
      "32596    68764888  68781469         41 68774709 68774750\n",
      "6720     68764888  68781469         41 68774884 68774925\n",
      "7477     68764888  68781469         41 68775083 68775124\n",
      "8014     68764888  68781469         41 68775686 68775727\n",
      "14633    68764888  68781469         41 68775733 68775774\n",
      "11541    68764888  68781469         41 68775821 68775862\n",
      "21776    68764888  68781469         43 68775854 68775897\n",
      "30367    68764888  68781469         41 68775920 68775961\n",
      "2146     68764888  68781469         41 68776000 68776041\n",
      "16194    68764888  68781469         41 68776047 68776088\n",
      "31868    68764888  68781469         41 68776182 68776223\n",
      "2113     68764888  68781469         41 68776247 68776288\n",
      "16195    68764888  68781469         41 68776294 68776335\n",
      "27632    68764888  68781469         43 68776419 68776462\n",
      "29030    68764888  68781469         41 68776484 68776525\n",
      "21198    68764888  68781469         43 68776592 68776635\n",
      "32981    68764888  68781469         41 68776657 68776698\n",
      "2006     68764888  68781469         41 68777856 68777897\n",
      "17431    68764888  68781469         41 68777903 68777944\n",
      "5928     68764888  68781469         41 68777991 68778032\n",
      "12321    68764888  68781469         41 68778528 68778569\n",
      "16870    68764888  68781469         41 68778575 68778616\n",
      "6546     68764888  68781469         41 68778663 68778704\n",
      "26747    68764888  68781469         43 68778696 68778739\n",
      "30297    68764888  68781469         41 68778761 68778802\n",
      "2548     68764888  68781469         41 68778843 68778884\n",
      "17392    68764888  68781469         41 68778890 68778931\n",
      "9245     68764888  68781469         41 68778978 68779019\n",
      "25183    68764888  68781469         43 68779011 68779054\n",
      "28938    68764888  68781469         41 68779077 68779118\n",
      "12258    68764888  68781469         41 68779221 68779262\n",
      "34309    68764888  68781469         50 68779230 68779280\n",
      "33267    68764888  68781469         41 68780173 68780214\n",
      "8519     68764888  68781469         41 68780271 68780312\n",
      "34447    68764888  68781469         50 68780280 68780330\n",
      "4017     68764888  68781469         41 68780410 68780451\n",
      "31729    68764888  68781469         41 68780491 68780532\n",
      "22466    68764888  68781469         43 68780555 68780598\n",
      "7341     68764888  68781469         41 68780590 68780631\n",
      "34981    68764888  68781469         50 68780599 68780649\n"
     ]
    }
   ],
   "source": [
    "filedir=paste0('sea_out_shrtn_',fprefix)\n",
    "sea_res_processed <- read.delim(paste0(filedir,'/sea_res_processed_',fprefix,'.tsv'))\n",
    "sea_res_problematic_seq <- sea_res_processed[sea_res_processed$seq_name %in% seq_names[problematic_seq_idx],]\n",
    "print(sea_res_problematic_seq[order(sea_res_problematic_seq$site_start),-c(4:7)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "db30b795",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of UTRs with motif cluster found: 3624 \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>combination</th><th scope=col>seq_name</th><th scope=col>chr</th><th scope=col>strand</th><th scope=col>meme_start_site_start</th><th scope=col>meme_end_site_start</th><th scope=col>meme_seq</th><th scope=col>ref_motif</th><th scope=col>min_consec_gap</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>TU11334_UTR3934</th><td>Orig_data                            </td><td>TU11334_UTR3934</td><td>chr5</td><td>-</td><td>4405</td><td>4634</td><td>MEME-1,MEME-2,MEME-4,MEME-3,MEME-5</td><td>MEME-1,MEME-2,MEME-4,MEME-3,MEME-5</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU12182_UTR4672</th><td>Combination_MEME-2_19553             </td><td>TU12182_UTR4672</td><td>chr6</td><td>+</td><td>3161</td><td>3393</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU15081_UTR6910</th><td>Orig_data                            </td><td>TU15081_UTR6910</td><td>chr7</td><td>-</td><td> 827</td><td>1059</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU16447_UTR7950.1</th><td>Combination_MEME-1_7238_MEME-1_12575 </td><td>TU16447_UTR7950</td><td>chr8</td><td>-</td><td> 238</td><td> 468</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU16447_UTR7950.2</th><td>Combination_MEME-3_25394_MEME-1_12575</td><td>TU16447_UTR7950</td><td>chr8</td><td>-</td><td> 238</td><td> 468</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>TU16447_UTR7950.3</th><td>Combination_MEME-4_34833_MEME-1_12575</td><td>TU16447_UTR7950</td><td>chr8</td><td>-</td><td> 238</td><td> 468</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>MEME-5,MEME-3,MEME-4,MEME-2,MEME-1</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & combination & seq\\_name & chr & strand & meme\\_start\\_site\\_start & meme\\_end\\_site\\_start & meme\\_seq & ref\\_motif & min\\_consec\\_gap\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\tTU11334\\_UTR3934 & Orig\\_data                             & TU11334\\_UTR3934 & chr5 & - & 4405 & 4634 & MEME-1,MEME-2,MEME-4,MEME-3,MEME-5 & MEME-1,MEME-2,MEME-4,MEME-3,MEME-5 & 0\\\\\n",
       "\tTU12182\\_UTR4672 & Combination\\_MEME-2\\_19553              & TU12182\\_UTR4672 & chr6 & + & 3161 & 3393 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & 0\\\\\n",
       "\tTU15081\\_UTR6910 & Orig\\_data                             & TU15081\\_UTR6910 & chr7 & - &  827 & 1059 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & 0\\\\\n",
       "\tTU16447\\_UTR7950.1 & Combination\\_MEME-1\\_7238\\_MEME-1\\_12575  & TU16447\\_UTR7950 & chr8 & - &  238 &  468 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & 0\\\\\n",
       "\tTU16447\\_UTR7950.2 & Combination\\_MEME-3\\_25394\\_MEME-1\\_12575 & TU16447\\_UTR7950 & chr8 & - &  238 &  468 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & 0\\\\\n",
       "\tTU16447\\_UTR7950.3 & Combination\\_MEME-4\\_34833\\_MEME-1\\_12575 & TU16447\\_UTR7950 & chr8 & - &  238 &  468 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | combination &lt;chr&gt; | seq_name &lt;chr&gt; | chr &lt;chr&gt; | strand &lt;chr&gt; | meme_start_site_start &lt;int&gt; | meme_end_site_start &lt;int&gt; | meme_seq &lt;chr&gt; | ref_motif &lt;chr&gt; | min_consec_gap &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| TU11334_UTR3934 | Orig_data                             | TU11334_UTR3934 | chr5 | - | 4405 | 4634 | MEME-1,MEME-2,MEME-4,MEME-3,MEME-5 | MEME-1,MEME-2,MEME-4,MEME-3,MEME-5 | 0 |\n",
       "| TU12182_UTR4672 | Combination_MEME-2_19553              | TU12182_UTR4672 | chr6 | + | 3161 | 3393 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | 0 |\n",
       "| TU15081_UTR6910 | Orig_data                             | TU15081_UTR6910 | chr7 | - |  827 | 1059 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | 0 |\n",
       "| TU16447_UTR7950.1 | Combination_MEME-1_7238_MEME-1_12575  | TU16447_UTR7950 | chr8 | - |  238 |  468 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | 0 |\n",
       "| TU16447_UTR7950.2 | Combination_MEME-3_25394_MEME-1_12575 | TU16447_UTR7950 | chr8 | - |  238 |  468 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | 0 |\n",
       "| TU16447_UTR7950.3 | Combination_MEME-4_34833_MEME-1_12575 | TU16447_UTR7950 | chr8 | - |  238 |  468 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "                  combination                           seq_name        chr \n",
       "TU11334_UTR3934   Orig_data                             TU11334_UTR3934 chr5\n",
       "TU12182_UTR4672   Combination_MEME-2_19553              TU12182_UTR4672 chr6\n",
       "TU15081_UTR6910   Orig_data                             TU15081_UTR6910 chr7\n",
       "TU16447_UTR7950.1 Combination_MEME-1_7238_MEME-1_12575  TU16447_UTR7950 chr8\n",
       "TU16447_UTR7950.2 Combination_MEME-3_25394_MEME-1_12575 TU16447_UTR7950 chr8\n",
       "TU16447_UTR7950.3 Combination_MEME-4_34833_MEME-1_12575 TU16447_UTR7950 chr8\n",
       "                  strand meme_start_site_start meme_end_site_start\n",
       "TU11334_UTR3934   -      4405                  4634               \n",
       "TU12182_UTR4672   +      3161                  3393               \n",
       "TU15081_UTR6910   -       827                  1059               \n",
       "TU16447_UTR7950.1 -       238                   468               \n",
       "TU16447_UTR7950.2 -       238                   468               \n",
       "TU16447_UTR7950.3 -       238                   468               \n",
       "                  meme_seq                          \n",
       "TU11334_UTR3934   MEME-1,MEME-2,MEME-4,MEME-3,MEME-5\n",
       "TU12182_UTR4672   MEME-5,MEME-3,MEME-4,MEME-2,MEME-1\n",
       "TU15081_UTR6910   MEME-5,MEME-3,MEME-4,MEME-2,MEME-1\n",
       "TU16447_UTR7950.1 MEME-5,MEME-3,MEME-4,MEME-2,MEME-1\n",
       "TU16447_UTR7950.2 MEME-5,MEME-3,MEME-4,MEME-2,MEME-1\n",
       "TU16447_UTR7950.3 MEME-5,MEME-3,MEME-4,MEME-2,MEME-1\n",
       "                  ref_motif                          min_consec_gap\n",
       "TU11334_UTR3934   MEME-1,MEME-2,MEME-4,MEME-3,MEME-5 0             \n",
       "TU12182_UTR4672   MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 0             \n",
       "TU15081_UTR6910   MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 0             \n",
       "TU16447_UTR7950.1 MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 0             \n",
       "TU16447_UTR7950.2 MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 0             \n",
       "TU16447_UTR7950.3 MEME-5,MEME-3,MEME-4,MEME-2,MEME-1 0             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filedir=paste0('sea_out_shrtn_',fprefix)\n",
    "sea_clst_found <- readRDS(file = paste0(filedir,'/sea_clst_found_',fprefix,'.rds'))\n",
    "sea_clst_found_df <- do.call('rbind',sea_clst_found)\n",
    "cat(\"Number of UTRs with motif cluster found:\", length(unique(sea_clst_found_df$seq_name)), \"\\n\")\n",
    "head(sea_clst_found_df[order(sea_clst_found_df$min_consec_gap),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1fe11228",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sea_clst_found_genes <- unique(gsub('_UTR[0-9]+','',sea_clst_found_df$seq_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a0217826",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TUs with full or partial motif cluster from discover set:  29"
     ]
    }
   ],
   "source": [
    "turef_shtn_mc_genes <- read.delim('turef_shtn_mc_genes.txt')\n",
    "cat('Number of TUs with full or partial motif cluster from discover set: ',length(unique(turef_shtn_mc_genes$tu)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7d614a60",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TU1169'</li><li>'TU26512'</li><li>'TU3179'</li><li>'TU34842'</li><li>'TU5795'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TU1169'\n",
       "\\item 'TU26512'\n",
       "\\item 'TU3179'\n",
       "\\item 'TU34842'\n",
       "\\item 'TU5795'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TU1169'\n",
       "2. 'TU26512'\n",
       "3. 'TU3179'\n",
       "4. 'TU34842'\n",
       "5. 'TU5795'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"TU1169\"  \"TU26512\" \"TU3179\"  \"TU34842\" \"TU5795\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setdiff(unique(turef_shtn_mc_genes$tu), sea_clst_found_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2d38ba5",
   "metadata": {},
   "source": [
    "\n",
    "Full or partial motif cluster false negatives Shortening: 'TU1169', 'TU26512', 'TU3179', 'TU34842', 'TU5795' (none were picked up even when relax min_nmotif to 2)\n",
    "\n",
    "Full or partial motif cluster fall negatives because cluster first detected in flanks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22ef65eb",
   "metadata": {},
   "source": [
    "#### Shortening motif cluster cleaning up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ac0ec81f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "calculate_overlap <- function(start1, end1, start2, end2) {\n",
    "  overlap_start <- max(start1, start2)\n",
    "  overlap_end <- min(end1, end2)\n",
    "  overlap_length <- max(0, overlap_end - overlap_start + 1)\n",
    "  \n",
    "  range1_length <- end1 - start1 + 1\n",
    "  range2_length <- end2 - start2 + 1\n",
    "  \n",
    "  # Calculate overlap as percentage of the smaller range\n",
    "  min_length <- min(range1_length, range2_length)\n",
    "  overlap_pct <- overlap_length / min_length\n",
    "  \n",
    "  return(overlap_pct)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4d86f37d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>tu</th><th scope=col>gene.id</th><th scope=col>name</th><th scope=col>utr</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>  70006</td><td>  71585</td><td>1580</td><td>+</td><td>TU4 </td><td>ENSG00000186092</td><td>OR4F5  </td><td>UTR1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td> 944151</td><td> 944575</td><td> 425</td><td>+</td><td>TU17</td><td>ENSG00000187634</td><td>SAMD11 </td><td>UTR2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td> 965189</td><td> 965719</td><td> 531</td><td>+</td><td>TU18</td><td>ENSG00000187961</td><td>KLHL17 </td><td>UTR3</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td> 974573</td><td> 975865</td><td>1293</td><td>+</td><td>TU19</td><td>ENSG00000187583</td><td>PLEKHN1</td><td>UTR4</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>1014476</td><td>1014540</td><td>  65</td><td>+</td><td>TU20</td><td>ENSG00000187608</td><td>ISG15  </td><td>UTR5</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>1054979</td><td>1056118</td><td>1140</td><td>+</td><td>TU21</td><td>ENSG00000188157</td><td>AGRN   </td><td>UTR6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & seqnames & start & end & width & strand & tu & gene.id & name & utr\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 &   70006 &   71585 & 1580 & + & TU4  & ENSG00000186092 & OR4F5   & UTR1\\\\\n",
       "\t2 & chr1 &  944151 &  944575 &  425 & + & TU17 & ENSG00000187634 & SAMD11  & UTR2\\\\\n",
       "\t3 & chr1 &  965189 &  965719 &  531 & + & TU18 & ENSG00000187961 & KLHL17  & UTR3\\\\\n",
       "\t4 & chr1 &  974573 &  975865 & 1293 & + & TU19 & ENSG00000187583 & PLEKHN1 & UTR4\\\\\n",
       "\t5 & chr1 & 1014476 & 1014540 &   65 & + & TU20 & ENSG00000187608 & ISG15   & UTR5\\\\\n",
       "\t6 & chr1 & 1054979 & 1056118 & 1140 & + & TU21 & ENSG00000188157 & AGRN    & UTR6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | tu &lt;chr&gt; | gene.id &lt;chr&gt; | name &lt;chr&gt; | utr &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 |   70006 |   71585 | 1580 | + | TU4  | ENSG00000186092 | OR4F5   | UTR1 |\n",
       "| 2 | chr1 |  944151 |  944575 |  425 | + | TU17 | ENSG00000187634 | SAMD11  | UTR2 |\n",
       "| 3 | chr1 |  965189 |  965719 |  531 | + | TU18 | ENSG00000187961 | KLHL17  | UTR3 |\n",
       "| 4 | chr1 |  974573 |  975865 | 1293 | + | TU19 | ENSG00000187583 | PLEKHN1 | UTR4 |\n",
       "| 5 | chr1 | 1014476 | 1014540 |   65 | + | TU20 | ENSG00000187608 | ISG15   | UTR5 |\n",
       "| 6 | chr1 | 1054979 | 1056118 | 1140 | + | TU21 | ENSG00000188157 | AGRN    | UTR6 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start   end     width strand tu   gene.id         name    utr \n",
       "1 chr1       70006   71585 1580  +      TU4  ENSG00000186092 OR4F5   UTR1\n",
       "2 chr1      944151  944575  425  +      TU17 ENSG00000187634 SAMD11  UTR2\n",
       "3 chr1      965189  965719  531  +      TU18 ENSG00000187961 KLHL17  UTR3\n",
       "4 chr1      974573  975865 1293  +      TU19 ENSG00000187583 PLEKHN1 UTR4\n",
       "5 chr1     1014476 1014540   65  +      TU20 ENSG00000187608 ISG15   UTR5\n",
       "6 chr1     1054979 1056118 1140  +      TU21 ENSG00000188157 AGRN    UTR6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes <- readRDS(\"/Users/ninhle/Desktop/Research/mcast_analyses/genes.rds\")\n",
    "ref <- genes$utr3 %>% as.data.frame()\n",
    "head(ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "912fc6cf",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 16</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seq_name</th><th scope=col>site_strand</th><th scope=col>site_start</th><th scope=col>pval</th><th scope=col>flank1</th><th scope=col>site</th><th scope=col>flank2</th><th scope=col>meme_name</th><th scope=col>strand</th><th scope=col>cis</th><th scope=col>chr</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>site_width</th><th scope=col>start</th><th scope=col>end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>TU3972_UTR23332 </td><td>-</td><td>1579</td><td>NA</td><td>NA</td><td>GATCTGCCTGCCTCGGCCTCCCAAAGTGCTGGGCTTACAAG</td><td>NA</td><td>MEME-1</td><td>+</td><td>FALSE</td><td>chr2 </td><td> 69879002</td><td> 69881385</td><td>41</td><td> 69880580</td><td> 69880621</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>TU4470_UTR23681 </td><td>+</td><td>2122</td><td>NA</td><td>NA</td><td>GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG</td><td>NA</td><td>MEME-1</td><td>+</td><td> TRUE</td><td>chr2 </td><td>167870018</td><td>167874046</td><td>41</td><td>167872139</td><td>167872180</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>TU27298_UTR16133</td><td>-</td><td> 890</td><td>NA</td><td>NA</td><td>GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG</td><td>NA</td><td>MEME-1</td><td>-</td><td> TRUE</td><td>chr15</td><td> 52115100</td><td> 52122760</td><td>41</td><td> 52121830</td><td> 52121871</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>TU483_UTR400    </td><td>+</td><td> 759</td><td>NA</td><td>NA</td><td>GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG</td><td>NA</td><td>MEME-1</td><td>+</td><td> TRUE</td><td>chr1 </td><td> 42798223</td><td> 42800828</td><td>41</td><td> 42798981</td><td> 42799022</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>TU18362_UTR9436 </td><td>-</td><td>4039</td><td>NA</td><td>NA</td><td>GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG</td><td>NA</td><td>MEME-1</td><td>-</td><td> TRUE</td><td>chr9 </td><td>134903232</td><td>134909801</td><td>41</td><td>134905722</td><td>134905763</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>TU19025_UTR9914 </td><td>+</td><td> 733</td><td>NA</td><td>NA</td><td>GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG</td><td>NA</td><td>MEME-1</td><td>+</td><td> TRUE</td><td>chr10</td><td>102900698</td><td>102901900</td><td>41</td><td>102901430</td><td>102901471</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 16\n",
       "\\begin{tabular}{r|llllllllllllllll}\n",
       "  & seq\\_name & site\\_strand & site\\_start & pval & flank1 & site & flank2 & meme\\_name & strand & cis & chr & range\\_start & range\\_end & site\\_width & start & end\\\\\n",
       "  & <chr> & <chr> & <int> & <lgl> & <lgl> & <chr> & <lgl> & <chr> & <chr> & <lgl> & <chr> & <int> & <int> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & TU3972\\_UTR23332  & - & 1579 & NA & NA & GATCTGCCTGCCTCGGCCTCCCAAAGTGCTGGGCTTACAAG & NA & MEME-1 & + & FALSE & chr2  &  69879002 &  69881385 & 41 &  69880580 &  69880621\\\\\n",
       "\t2 & TU4470\\_UTR23681  & + & 2122 & NA & NA & GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG & NA & MEME-1 & + &  TRUE & chr2  & 167870018 & 167874046 & 41 & 167872139 & 167872180\\\\\n",
       "\t3 & TU27298\\_UTR16133 & - &  890 & NA & NA & GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG & NA & MEME-1 & - &  TRUE & chr15 &  52115100 &  52122760 & 41 &  52121830 &  52121871\\\\\n",
       "\t4 & TU483\\_UTR400     & + &  759 & NA & NA & GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG & NA & MEME-1 & + &  TRUE & chr1  &  42798223 &  42800828 & 41 &  42798981 &  42799022\\\\\n",
       "\t5 & TU18362\\_UTR9436  & - & 4039 & NA & NA & GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG & NA & MEME-1 & - &  TRUE & chr9  & 134903232 & 134909801 & 41 & 134905722 & 134905763\\\\\n",
       "\t6 & TU19025\\_UTR9914  & + &  733 & NA & NA & GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG & NA & MEME-1 & + &  TRUE & chr10 & 102900698 & 102901900 & 41 & 102901430 & 102901471\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 16\n",
       "\n",
       "| <!--/--> | seq_name &lt;chr&gt; | site_strand &lt;chr&gt; | site_start &lt;int&gt; | pval &lt;lgl&gt; | flank1 &lt;lgl&gt; | site &lt;chr&gt; | flank2 &lt;lgl&gt; | meme_name &lt;chr&gt; | strand &lt;chr&gt; | cis &lt;lgl&gt; | chr &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | site_width &lt;int&gt; | start &lt;int&gt; | end &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | TU3972_UTR23332  | - | 1579 | NA | NA | GATCTGCCTGCCTCGGCCTCCCAAAGTGCTGGGCTTACAAG | NA | MEME-1 | + | FALSE | chr2  |  69879002 |  69881385 | 41 |  69880580 |  69880621 |\n",
       "| 2 | TU4470_UTR23681  | + | 2122 | NA | NA | GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG | NA | MEME-1 | + |  TRUE | chr2  | 167870018 | 167874046 | 41 | 167872139 | 167872180 |\n",
       "| 3 | TU27298_UTR16133 | - |  890 | NA | NA | GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG | NA | MEME-1 | - |  TRUE | chr15 |  52115100 |  52122760 | 41 |  52121830 |  52121871 |\n",
       "| 4 | TU483_UTR400     | + |  759 | NA | NA | GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG | NA | MEME-1 | + |  TRUE | chr1  |  42798223 |  42800828 | 41 |  42798981 |  42799022 |\n",
       "| 5 | TU18362_UTR9436  | - | 4039 | NA | NA | GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG | NA | MEME-1 | - |  TRUE | chr9  | 134903232 | 134909801 | 41 | 134905722 | 134905763 |\n",
       "| 6 | TU19025_UTR9914  | + |  733 | NA | NA | GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG | NA | MEME-1 | + |  TRUE | chr10 | 102900698 | 102901900 | 41 | 102901430 | 102901471 |\n",
       "\n"
      ],
      "text/plain": [
       "  seq_name         site_strand site_start pval flank1\n",
       "1 TU3972_UTR23332  -           1579       NA   NA    \n",
       "2 TU4470_UTR23681  +           2122       NA   NA    \n",
       "3 TU27298_UTR16133 -            890       NA   NA    \n",
       "4 TU483_UTR400     +            759       NA   NA    \n",
       "5 TU18362_UTR9436  -           4039       NA   NA    \n",
       "6 TU19025_UTR9914  +            733       NA   NA    \n",
       "  site                                      flank2 meme_name strand cis   chr  \n",
       "1 GATCTGCCTGCCTCGGCCTCCCAAAGTGCTGGGCTTACAAG NA     MEME-1    +      FALSE chr2 \n",
       "2 GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG NA     MEME-1    +       TRUE chr2 \n",
       "3 GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG NA     MEME-1    -       TRUE chr15\n",
       "4 GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG NA     MEME-1    +       TRUE chr1 \n",
       "5 GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG NA     MEME-1    -       TRUE chr9 \n",
       "6 GATCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAAG NA     MEME-1    +       TRUE chr10\n",
       "  range_start range_end site_width start     end      \n",
       "1  69879002    69881385 41          69880580  69880621\n",
       "2 167870018   167874046 41         167872139 167872180\n",
       "3  52115100    52122760 41          52121830  52121871\n",
       "4  42798223    42800828 41          42798981  42799022\n",
       "5 134903232   134909801 41         134905722 134905763\n",
       "6 102900698   102901900 41         102901430 102901471"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Group overlapping ranges within each seq_name\n",
    "filedir=paste0('sea_out_shrtn_',fprefix)\n",
    "sea_res <- read.delim(paste0(filedir,'/sea_res_processed_',fprefix,'.tsv'))\n",
    "head(sea_res)\n",
    "sea_clst_found <- readRDS(file = paste0(filedir,'/sea_clst_found_',fprefix,'.rds'))\n",
    "sea_clst_found_df <- do.call('rbind',sea_clst_found)\n",
    "sea_clst_grouped <- sea_clst_found_df %>%\n",
    "  group_by(seq_name) %>%\n",
    "  arrange(meme_start_site_start) %>%\n",
    "  mutate(\n",
    "    range_length = meme_end_site_start - meme_start_site_start + 1,\n",
    "    overlap_group = {\n",
    "      n_rows <- n()\n",
    "      if (n_rows == 1) {\n",
    "        1\n",
    "      } else {\n",
    "        groups <- rep(NA, n_rows)\n",
    "        groups[1] <- 1\n",
    "        current_group <- 1\n",
    "        \n",
    "        for (i in 2:n_rows) {\n",
    "          # Check overlap with previous rows in current group\n",
    "          overlaps_with_group <- FALSE\n",
    "          for (j in 1:(i-1)) {\n",
    "            if (groups[j] == current_group) {\n",
    "              overlap_pct <- calculate_overlap(\n",
    "                meme_start_site_start[i], meme_end_site_start[i],\n",
    "                meme_start_site_start[j], meme_end_site_start[j]\n",
    "              )\n",
    "              if (overlap_pct >= 0.9) {\n",
    "                overlaps_with_group <- TRUE\n",
    "                break\n",
    "              }\n",
    "            }\n",
    "          }\n",
    "          \n",
    "          if (overlaps_with_group) {\n",
    "            groups[i] <- current_group\n",
    "          } else {\n",
    "            current_group <- current_group + 1\n",
    "            groups[i] <- current_group\n",
    "          }\n",
    "        }\n",
    "        groups\n",
    "      }\n",
    "    }\n",
    "  ) %>%\n",
    "  group_by(seq_name, overlap_group) %>%\n",
    "  summarize(\n",
    "    cluster_count = n(),\n",
    "    all_combinations = paste(unique(combination), collapse = \";\"),\n",
    "    all_meme_seqs = paste(unique(meme_seq), collapse = \";\"),\n",
    "    all_starts = {\n",
    "    unique_seqs <- unique(meme_seq)\n",
    "    start_values <- sapply(unique_seqs, function(seq) meme_start_site_start[meme_seq == seq][1])\n",
    "    list(start_values)},\n",
    "    all_ends = {\n",
    "    unique_seqs <- unique(meme_seq)\n",
    "    end_values <- sapply(unique_seqs, function(seq) meme_end_site_start[meme_seq == seq][1])\n",
    "    list(end_values)},\n",
    "    .groups = \"drop\"\n",
    "  ) %>%\n",
    "  rowwise() %>%\n",
    "  mutate(\n",
    "    # Count motifs in each meme_seq\n",
    "    meme_seqs_list = list(strsplit(all_meme_seqs, \";\")[[1]]),\n",
    "    motif_counts = list(sapply(meme_seqs_list, function(x) length(strsplit(x, \",\")[[1]]))),\n",
    "    max_motif_count = max(unlist(motif_counts)),\n",
    "    \n",
    "    # Find meme_seqs with maximum motif count\n",
    "    max_meme_seqs = list(meme_seqs_list[motif_counts == max_motif_count]),\n",
    "  \n",
    "    # Extract final values directly\n",
    "    meme_seqs = if(length(max_meme_seqs) == 1) {\n",
    "    max_meme_seqs\n",
    "    } else {\n",
    "      paste(max_meme_seqs, collapse = \";\")\n",
    "    },\n",
    "    full_motif_cluster = max_motif_count == 5,\n",
    "\n",
    "    min_start = if(length(max_meme_seqs) == 1) {\n",
    "      max_seq = max_meme_seqs\n",
    "      seq_idx = which(meme_seqs_list == max_seq)\n",
    "      all_starts[seq_idx]\n",
    "    } else {\n",
    "      max_indices = which(motif_counts == max_motif_count)\n",
    "      min(all_starts[max_indices])},\n",
    "    \n",
    "    max_end = if(length(max_meme_seqs) == 1) {\n",
    "      max_seq = max_meme_seqs\n",
    "      seq_idx = which(meme_seqs_list == max_seq)\n",
    "      all_ends[seq_idx]\n",
    "    } else {\n",
    "      max_indices = which(motif_counts == max_motif_count)\n",
    "      max(all_ends[max_indices])\n",
    "    },\n",
    "  \n",
    "  combinations = all_combinations\n",
    ") %>%\n",
    "  ungroup() %>%\n",
    "  select(seq_name, overlap_group, cluster_count, full_motif_cluster, min_start, max_end, combinations, meme_seqs) %>%\n",
    "  arrange(seq_name, min_start)\n",
    "sea_clst_grouped$tu <- gsub('_UTR[0-9]+','',sea_clst_grouped$seq_name)\n",
    "sea_clst_grouped$gene <- ref$name[match(sea_clst_grouped$tu,ref$tu)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3a20a861",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sea_clst_grouped$strand <- seq_bed$strand[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "sea_clst_grouped$chr <- seq_bed$chr[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "sea_clst_grouped$range_start <- seq_bed$start[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "sea_clst_grouped$range_end <- seq_bed$end[match(sea_clst_grouped$seq_name,seq_bed$seq_name)]\n",
    "\n",
    "\n",
    "sea_clst_grouped$most_upstream_motif_start <- apply(sea_clst_grouped,1,function(x){\n",
    "  strand <- x['strand']\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  min_start <- as.numeric(x['min_start'])\n",
    "  chr <- x['chr']\n",
    "  start <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==min_start),'start'] \n",
    "  return(start)\n",
    "})\n",
    "\n",
    "sea_clst_grouped$most_upstream_motif <- apply(sea_clst_grouped,1,function(x){\n",
    "  strand <- x['strand']\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  min_start <- as.numeric(x['min_start'])\n",
    "  chr <- x['chr']\n",
    "  motif <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==min_start),'meme_name']\n",
    "  return(motif)\n",
    "})\n",
    "\n",
    "sea_clst_grouped$most_downstream_motif_start <- apply(sea_clst_grouped,1,function(x){\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  max_end <- as.numeric(x['max_end'])\n",
    "  chr <- x['chr']\n",
    "  start <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==max_end),'start']\n",
    "  return(start)\n",
    "})\n",
    "\n",
    "sea_clst_grouped$most_downstream_motif <- apply(sea_clst_grouped,1,function(x){\n",
    "  seq_name <- as.character(x['seq_name'])\n",
    "  max_end <- as.numeric(x['max_end'])\n",
    "  chr <- x['chr']\n",
    "  motif <- sea_res[which(sea_res$seq_name==seq_name & sea_res$chr==chr & sea_res$site_start==max_end),'meme_name']\n",
    "  return(motif)\n",
    "})\n",
    "\n",
    "\n",
    "sea_clst_grouped <- sea_clst_grouped %>%\n",
    "  relocate(chr, .after = seq_name) %>%\n",
    "  relocate(strand, .after = chr) %>%\n",
    "  relocate(range_start, .after = strand) %>%\n",
    "  relocate(range_end, .after = range_start) %>%\n",
    "  relocate(meme_seqs, .after = range_end) %>%\n",
    "  relocate(full_motif_cluster, .after = meme_seqs) %>%\n",
    "  relocate(most_upstream_motif, .after = full_motif_cluster) %>%\n",
    "  relocate(most_upstream_motif_start, .after = most_upstream_motif) %>%\n",
    "  relocate(most_downstream_motif, .after = most_upstream_motif_start) %>%\n",
    "  relocate(most_downstream_motif_start, .after = most_downstream_motif) %>%\n",
    "  relocate(cluster_count, .after = most_downstream_motif_start) %>%\n",
    "  relocate(tu, .before = seq_name) %>% \n",
    "  relocate(gene, .before = tu) %>% select(-c(combinations,cluster_count,overlap_group,min_start,max_end))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "49caef40",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 6 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene</th><th scope=col>tu</th><th scope=col>seq_name</th><th scope=col>chr</th><th scope=col>strand</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>meme_seqs</th><th scope=col>full_motif_cluster</th><th scope=col>most_upstream_motif</th><th scope=col>most_upstream_motif_start</th><th scope=col>most_downstream_motif</th><th scope=col>most_downstream_motif_start</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG </td><td>TU34454</td><td>TU34454_UTR0</td><td>chr19</td><td>-</td><td> 58345183</td><td> 58347025</td><td>MEME-5,MEME-3,MEME-1                                   </td><td>FALSE</td><td>MEME-5</td><td> 58345641</td><td>MEME-1</td><td> 58345403</td></tr>\n",
       "\t<tr><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR1</td><td>chr12</td><td>+</td><td>  8875461</td><td>  8875815</td><td>MEME-3,MEME-2,MEME-1                                   </td><td>FALSE</td><td>MEME-3</td><td>  8875503</td><td>MEME-1</td><td>  8875676</td></tr>\n",
       "\t<tr><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR2</td><td>chr12</td><td>+</td><td>  8876058</td><td>  8876788</td><td>MEME-1,MEME-2,MEME-3                                   </td><td>FALSE</td><td>MEME-1</td><td>  8876412</td><td>MEME-3</td><td>  8876579</td></tr>\n",
       "\t<tr><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR1 </td><td>chr2 </td><td>-</td><td> 69481056</td><td> 69482589</td><td>MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1</td><td>FALSE</td><td>MEME-5</td><td> 69482325</td><td>MEME-1</td><td> 69482091</td></tr>\n",
       "\t<tr><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR2 </td><td>chr2 </td><td>-</td><td> 69457997</td><td> 69475872</td><td>MEME-5,MEME-3,MEME-2,MEME-1                            </td><td>FALSE</td><td>MEME-5</td><td> 69462648</td><td>MEME-1</td><td> 69462420</td></tr>\n",
       "\t<tr><td>AARD </td><td>TU16086</td><td>TU16086_UTR0</td><td>chr8 </td><td>+</td><td>116942699</td><td>116944488</td><td>MEME-1,MEME-2,MEME-3,MEME-5                            </td><td>FALSE</td><td>MEME-1</td><td>116942739</td><td>MEME-5</td><td>116942971</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 6 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " gene & tu & seq\\_name & chr & strand & range\\_start & range\\_end & meme\\_seqs & full\\_motif\\_cluster & most\\_upstream\\_motif & most\\_upstream\\_motif\\_start & most\\_downstream\\_motif & most\\_downstream\\_motif\\_start\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <lgl> & <chr> & <int> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t A1BG  & TU34454 & TU34454\\_UTR0 & chr19 & - &  58345183 &  58347025 & MEME-5,MEME-3,MEME-1                                    & FALSE & MEME-5 &  58345641 & MEME-1 &  58345403\\\\\n",
       "\t A2ML1 & TU22196 & TU22196\\_UTR1 & chr12 & + &   8875461 &   8875815 & MEME-3,MEME-2,MEME-1                                    & FALSE & MEME-3 &   8875503 & MEME-1 &   8875676\\\\\n",
       "\t A2ML1 & TU22196 & TU22196\\_UTR2 & chr12 & + &   8876058 &   8876788 & MEME-1,MEME-2,MEME-3                                    & FALSE & MEME-1 &   8876412 & MEME-3 &   8876579\\\\\n",
       "\t AAK1  & TU5339  & TU5339\\_UTR1  & chr2  & - &  69481056 &  69482589 & MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1 & FALSE & MEME-5 &  69482325 & MEME-1 &  69482091\\\\\n",
       "\t AAK1  & TU5339  & TU5339\\_UTR2  & chr2  & - &  69457997 &  69475872 & MEME-5,MEME-3,MEME-2,MEME-1                             & FALSE & MEME-5 &  69462648 & MEME-1 &  69462420\\\\\n",
       "\t AARD  & TU16086 & TU16086\\_UTR0 & chr8  & + & 116942699 & 116944488 & MEME-1,MEME-2,MEME-3,MEME-5                             & FALSE & MEME-1 & 116942739 & MEME-5 & 116942971\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 6 × 13\n",
       "\n",
       "| gene &lt;chr&gt; | tu &lt;chr&gt; | seq_name &lt;chr&gt; | chr &lt;chr&gt; | strand &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | meme_seqs &lt;chr&gt; | full_motif_cluster &lt;lgl&gt; | most_upstream_motif &lt;chr&gt; | most_upstream_motif_start &lt;int&gt; | most_downstream_motif &lt;chr&gt; | most_downstream_motif_start &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A1BG  | TU34454 | TU34454_UTR0 | chr19 | - |  58345183 |  58347025 | MEME-5,MEME-3,MEME-1                                    | FALSE | MEME-5 |  58345641 | MEME-1 |  58345403 |\n",
       "| A2ML1 | TU22196 | TU22196_UTR1 | chr12 | + |   8875461 |   8875815 | MEME-3,MEME-2,MEME-1                                    | FALSE | MEME-3 |   8875503 | MEME-1 |   8875676 |\n",
       "| A2ML1 | TU22196 | TU22196_UTR2 | chr12 | + |   8876058 |   8876788 | MEME-1,MEME-2,MEME-3                                    | FALSE | MEME-1 |   8876412 | MEME-3 |   8876579 |\n",
       "| AAK1  | TU5339  | TU5339_UTR1  | chr2  | - |  69481056 |  69482589 | MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1 | FALSE | MEME-5 |  69482325 | MEME-1 |  69482091 |\n",
       "| AAK1  | TU5339  | TU5339_UTR2  | chr2  | - |  69457997 |  69475872 | MEME-5,MEME-3,MEME-2,MEME-1                             | FALSE | MEME-5 |  69462648 | MEME-1 |  69462420 |\n",
       "| AARD  | TU16086 | TU16086_UTR0 | chr8  | + | 116942699 | 116944488 | MEME-1,MEME-2,MEME-3,MEME-5                             | FALSE | MEME-1 | 116942739 | MEME-5 | 116942971 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene  tu      seq_name     chr   strand range_start range_end\n",
       "1 A1BG  TU34454 TU34454_UTR0 chr19 -       58345183    58347025\n",
       "2 A2ML1 TU22196 TU22196_UTR1 chr12 +        8875461     8875815\n",
       "3 A2ML1 TU22196 TU22196_UTR2 chr12 +        8876058     8876788\n",
       "4 AAK1  TU5339  TU5339_UTR1  chr2  -       69481056    69482589\n",
       "5 AAK1  TU5339  TU5339_UTR2  chr2  -       69457997    69475872\n",
       "6 AARD  TU16086 TU16086_UTR0 chr8  +      116942699   116944488\n",
       "  meme_seqs                                               full_motif_cluster\n",
       "1 MEME-5,MEME-3,MEME-1                                    FALSE             \n",
       "2 MEME-3,MEME-2,MEME-1                                    FALSE             \n",
       "3 MEME-1,MEME-2,MEME-3                                    FALSE             \n",
       "4 MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1 FALSE             \n",
       "5 MEME-5,MEME-3,MEME-2,MEME-1                             FALSE             \n",
       "6 MEME-1,MEME-2,MEME-3,MEME-5                             FALSE             \n",
       "  most_upstream_motif most_upstream_motif_start most_downstream_motif\n",
       "1 MEME-5               58345641                 MEME-1               \n",
       "2 MEME-3                8875503                 MEME-1               \n",
       "3 MEME-1                8876412                 MEME-3               \n",
       "4 MEME-5               69482325                 MEME-1               \n",
       "5 MEME-5               69462648                 MEME-1               \n",
       "6 MEME-1              116942739                 MEME-5               \n",
       "  most_downstream_motif_start\n",
       "1  58345403                  \n",
       "2   8875676                  \n",
       "3   8876579                  \n",
       "4  69482091                  \n",
       "5  69462420                  \n",
       "6 116942971                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sea_clst_grouped <- sea_clst_grouped %>%\n",
    "  # Extract TU and current UTR index for sorting\n",
    "  mutate(\n",
    "    tu_part = gsub('_UTR[0-9]+', '', seq_name),\n",
    "    current_utr = as.numeric(gsub('.*_UTR([0-9]+)', '\\\\1', seq_name))\n",
    "  ) %>%\n",
    "  group_by(tu_part) %>%\n",
    "  arrange(\n",
    "    tu_part, \n",
    "    case_when(\n",
    "      strand == '+' ~ range_start,\n",
    "      strand == '-' ~ -range_end\n",
    "    )\n",
    "  ) %>%\n",
    "  mutate(\n",
    "    # Create unique UTR identifier\n",
    "    unique_seqs = list(unique(seq_name)),\n",
    "    # Assign index \n",
    "    new_utr_index = case_when(\n",
    "      length(seq_name) == 1 ~ 0,\n",
    "      TRUE ~ match(seq_name, unique_seqs[[1]]) \n",
    "    ),\n",
    "    # Create new seq_name with reindexed UTR\n",
    "    seq_name = paste0(tu_part, '_UTR', new_utr_index)\n",
    "  ) %>%\n",
    "  ungroup() %>%\n",
    "  select(-tu_part, -current_utr, -new_utr_index, -unique_seqs) %>%\n",
    "  arrange(gene, tu, seq_name)\n",
    "\n",
    "head(sea_clst_grouped)\n",
    "write.table(sea_clst_grouped,\n",
    "            file = paste0(filedir,'/','motif_cluster_detection_res_',fprefix,'.tsv'),\n",
    "            sep = '\\t',quote = F,row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c3f52f5d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "3443"
      ],
      "text/latex": [
       "3443"
      ],
      "text/markdown": [
       "3443"
      ],
      "text/plain": [
       "[1] 3443"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Number of TUs with shortening motif clusters\n",
    "length(unique(sea_clst_grouped$tu))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f02d9785",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene</th><th scope=col>tu</th><th scope=col>seq_name</th><th scope=col>chr</th><th scope=col>strand</th><th scope=col>range_start</th><th scope=col>range_end</th><th scope=col>meme_seqs</th><th scope=col>full_motif_cluster</th><th scope=col>most_upstream_motif</th><th scope=col>most_upstream_motif_start</th><th scope=col>most_downstream_motif</th><th scope=col>most_downstream_motif_start</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG </td><td>TU34454</td><td>TU34454_UTR0</td><td>chr19</td><td>-</td><td> 58345183</td><td> 58347025</td><td>MEME-5,MEME-3,MEME-1                                   </td><td>FALSE</td><td>MEME-5</td><td> 58345641</td><td>MEME-1</td><td> 58345403</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR1</td><td>chr12</td><td>+</td><td>  8875461</td><td>  8875815</td><td>MEME-3,MEME-2,MEME-1                                   </td><td>FALSE</td><td>MEME-3</td><td>  8875503</td><td>MEME-1</td><td>  8875676</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>A2ML1</td><td>TU22196</td><td>TU22196_UTR2</td><td>chr12</td><td>+</td><td>  8876058</td><td>  8876788</td><td>MEME-1,MEME-2,MEME-3                                   </td><td>FALSE</td><td>MEME-1</td><td>  8876412</td><td>MEME-3</td><td>  8876579</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR1 </td><td>chr2 </td><td>-</td><td> 69481056</td><td> 69482589</td><td>MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1</td><td>FALSE</td><td>MEME-5</td><td> 69482325</td><td>MEME-1</td><td> 69482091</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>AAK1 </td><td>TU5339 </td><td>TU5339_UTR2 </td><td>chr2 </td><td>-</td><td> 69457997</td><td> 69475872</td><td>MEME-5,MEME-3,MEME-2,MEME-1                            </td><td>FALSE</td><td>MEME-5</td><td> 69462648</td><td>MEME-1</td><td> 69462420</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>AARD </td><td>TU16086</td><td>TU16086_UTR0</td><td>chr8 </td><td>+</td><td>116942699</td><td>116944488</td><td>MEME-1,MEME-2,MEME-3,MEME-5                            </td><td>FALSE</td><td>MEME-1</td><td>116942739</td><td>MEME-5</td><td>116942971</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 13\n",
       "\\begin{tabular}{r|lllllllllllll}\n",
       "  & gene & tu & seq\\_name & chr & strand & range\\_start & range\\_end & meme\\_seqs & full\\_motif\\_cluster & most\\_upstream\\_motif & most\\_upstream\\_motif\\_start & most\\_downstream\\_motif & most\\_downstream\\_motif\\_start\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <lgl> & <chr> & <int> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG  & TU34454 & TU34454\\_UTR0 & chr19 & - &  58345183 &  58347025 & MEME-5,MEME-3,MEME-1                                    & FALSE & MEME-5 &  58345641 & MEME-1 &  58345403\\\\\n",
       "\t2 & A2ML1 & TU22196 & TU22196\\_UTR1 & chr12 & + &   8875461 &   8875815 & MEME-3,MEME-2,MEME-1                                    & FALSE & MEME-3 &   8875503 & MEME-1 &   8875676\\\\\n",
       "\t3 & A2ML1 & TU22196 & TU22196\\_UTR2 & chr12 & + &   8876058 &   8876788 & MEME-1,MEME-2,MEME-3                                    & FALSE & MEME-1 &   8876412 & MEME-3 &   8876579\\\\\n",
       "\t4 & AAK1  & TU5339  & TU5339\\_UTR1  & chr2  & - &  69481056 &  69482589 & MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1 & FALSE & MEME-5 &  69482325 & MEME-1 &  69482091\\\\\n",
       "\t5 & AAK1  & TU5339  & TU5339\\_UTR2  & chr2  & - &  69457997 &  69475872 & MEME-5,MEME-3,MEME-2,MEME-1                             & FALSE & MEME-5 &  69462648 & MEME-1 &  69462420\\\\\n",
       "\t6 & AARD  & TU16086 & TU16086\\_UTR0 & chr8  & + & 116942699 & 116944488 & MEME-1,MEME-2,MEME-3,MEME-5                             & FALSE & MEME-1 & 116942739 & MEME-5 & 116942971\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 13\n",
       "\n",
       "| <!--/--> | gene &lt;chr&gt; | tu &lt;chr&gt; | seq_name &lt;chr&gt; | chr &lt;chr&gt; | strand &lt;chr&gt; | range_start &lt;int&gt; | range_end &lt;int&gt; | meme_seqs &lt;chr&gt; | full_motif_cluster &lt;lgl&gt; | most_upstream_motif &lt;chr&gt; | most_upstream_motif_start &lt;int&gt; | most_downstream_motif &lt;chr&gt; | most_downstream_motif_start &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A1BG  | TU34454 | TU34454_UTR0 | chr19 | - |  58345183 |  58347025 | MEME-5,MEME-3,MEME-1                                    | FALSE | MEME-5 |  58345641 | MEME-1 |  58345403 |\n",
       "| 2 | A2ML1 | TU22196 | TU22196_UTR1 | chr12 | + |   8875461 |   8875815 | MEME-3,MEME-2,MEME-1                                    | FALSE | MEME-3 |   8875503 | MEME-1 |   8875676 |\n",
       "| 3 | A2ML1 | TU22196 | TU22196_UTR2 | chr12 | + |   8876058 |   8876788 | MEME-1,MEME-2,MEME-3                                    | FALSE | MEME-1 |   8876412 | MEME-3 |   8876579 |\n",
       "| 4 | AAK1  | TU5339  | TU5339_UTR1  | chr2  | - |  69481056 |  69482589 | MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1 | FALSE | MEME-5 |  69482325 | MEME-1 |  69482091 |\n",
       "| 5 | AAK1  | TU5339  | TU5339_UTR2  | chr2  | - |  69457997 |  69475872 | MEME-5,MEME-3,MEME-2,MEME-1                             | FALSE | MEME-5 |  69462648 | MEME-1 |  69462420 |\n",
       "| 6 | AARD  | TU16086 | TU16086_UTR0 | chr8  | + | 116942699 | 116944488 | MEME-1,MEME-2,MEME-3,MEME-5                             | FALSE | MEME-1 | 116942739 | MEME-5 | 116942971 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene  tu      seq_name     chr   strand range_start range_end\n",
       "1 A1BG  TU34454 TU34454_UTR0 chr19 -       58345183    58347025\n",
       "2 A2ML1 TU22196 TU22196_UTR1 chr12 +        8875461     8875815\n",
       "3 A2ML1 TU22196 TU22196_UTR2 chr12 +        8876058     8876788\n",
       "4 AAK1  TU5339  TU5339_UTR1  chr2  -       69481056    69482589\n",
       "5 AAK1  TU5339  TU5339_UTR2  chr2  -       69457997    69475872\n",
       "6 AARD  TU16086 TU16086_UTR0 chr8  +      116942699   116944488\n",
       "  meme_seqs                                               full_motif_cluster\n",
       "1 MEME-5,MEME-3,MEME-1                                    FALSE             \n",
       "2 MEME-3,MEME-2,MEME-1                                    FALSE             \n",
       "3 MEME-1,MEME-2,MEME-3                                    FALSE             \n",
       "4 MEME-5,MEME-3,MEME-2,MEME-1;MEME-5,MEME-4,MEME-2,MEME-1 FALSE             \n",
       "5 MEME-5,MEME-3,MEME-2,MEME-1                             FALSE             \n",
       "6 MEME-1,MEME-2,MEME-3,MEME-5                             FALSE             \n",
       "  most_upstream_motif most_upstream_motif_start most_downstream_motif\n",
       "1 MEME-5               58345641                 MEME-1               \n",
       "2 MEME-3                8875503                 MEME-1               \n",
       "3 MEME-1                8876412                 MEME-3               \n",
       "4 MEME-5               69482325                 MEME-1               \n",
       "5 MEME-5               69462648                 MEME-1               \n",
       "6 MEME-1              116942739                 MEME-5               \n",
       "  most_downstream_motif_start\n",
       "1  58345403                  \n",
       "2   8875676                  \n",
       "3   8876579                  \n",
       "4  69482091                  \n",
       "5  69462420                  \n",
       "6 116942971                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of full or partial shortening motif cluster: 3722 \n",
      "Number of full or partial shortening motif cluster outside discovery set: 3695 \n",
      "Number of TUs with full or partial shortening motif cluster: 3443 \n",
      "Number of TUs with full or partial shortening motif cluster outside discovery set: 3419 \n"
     ]
    }
   ],
   "source": [
    "filedir=paste0('sea_out_shrtn_',fprefix)\n",
    "sea_clst_grouped <- read.delim(paste0(filedir,'/','motif_cluster_detection_res_', fprefix, '.tsv'))\n",
    "sea_clst_grouped_outside_discovery_set <- sea_clst_grouped[!(sea_clst_grouped$tu %in% turef_shtn_mc_genes$tu),]\n",
    "write.table(sea_clst_grouped_outside_discovery_set,\n",
    "            file = paste0(filedir,'/','motif_cluster_detection_res_outside_discovery_set_',fprefix,'.tsv'),\n",
    "            sep = '\\t',quote = F,row.names = F)\n",
    "head(sea_clst_grouped_outside_discovery_set)\n",
    "cat('Number of full or partial shortening motif cluster:', nrow(sea_clst_grouped), '\\n')\n",
    "cat('Number of full or partial shortening motif cluster outside discovery set:', nrow(sea_clst_grouped_outside_discovery_set), '\\n')\n",
    "cat('Number of TUs with full or partial shortening motif cluster:', length(unique(sea_clst_grouped$tu)), '\\n')\n",
    "cat('Number of TUs with full or partial shortening motif cluster outside discovery set:', length(unique(sea_clst_grouped_outside_discovery_set$tu)), '\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
