{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3c19d07",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: generics\n",
      "\n",
      "\n",
      "Attaching package: ‘generics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,\n",
      "    setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,\n",
      "    unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: data.table\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "Loading required package: stringr\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘IRanges::shift’ by ‘data.table::shift’ when loading ‘goldmine’”\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:generics’:\n",
      "\n",
      "    explain\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(goldmine)\n",
    "library(dplyr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f4064841",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "wd <- \"/Users/ninhle/Desktop/Research/scPASU_pipeline_runs/Ureter10_scPASU_run/outputs/differentiation_stage_cellranger_peakcount/\"\n",
    "setwd(wd)\n",
    "genes <- readRDS(\"/Users/ninhle/Desktop/Research/mcast_analyses/genes.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6f4e2db5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Check if any ranges overlap in a GRanges object\n",
    "check_overlaps <- function(gr) {\n",
    "  overlaps <- findOverlaps(gr, gr, ignore.strand = TRUE)\n",
    "  # Remove self-overlaps\n",
    "  overlaps <- overlaps[queryHits(overlaps) != subjectHits(overlaps)]\n",
    "  has_overlaps <- length(overlaps) > 0\n",
    "  \n",
    "  return(list(\n",
    "    has_overlaps = has_overlaps,\n",
    "    n_overlapping_pairs = length(overlaps),\n",
    "    overlapping_indices = overlaps\n",
    "  ))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "460b07a4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting overlap checking for 19359 transcript units at 2025-12-09 03:08:54 \n",
      "  |++++++++++++++++++++++++++++++++++++++++++++++++++| 100%\n",
      "Overlap checking completed! Total time: 112.03 seconds (1.87 minutes)\n"
     ]
    }
   ],
   "source": [
    "library(pbapply)\n",
    "\n",
    "# Sanity check to make sure no UTRs are overlapping within the same transcript unit\n",
    "unique_tus <- unique(genes$utr3$tu)\n",
    "start_time <- Sys.time()\n",
    "cat(\"Starting overlap checking for\", length(unique_tus), \"transcript units at\", format(start_time), \"\\n\")\n",
    "\n",
    "# Force progress bar to show\n",
    "pboptions(type = \"txt\")\n",
    "\n",
    "tu_ovl_results <- pblapply(unique_tus, function(tu) {\n",
    "  gr_subset <- genes$utr3[genes$utr3$tu == tu]\n",
    "  result <- check_overlaps(gr_subset)\n",
    "  result$tu <- tu\n",
    "  return(result)\n",
    "})\n",
    "\n",
    "names(tu_ovl_results) <- unique_tus\n",
    "\n",
    "total_time <- as.numeric(difftime(Sys.time(), start_time, units = \"secs\"))\n",
    "cat(sprintf(\"Overlap checking completed! Total time: %.2f seconds (%.2f minutes)\\n\", \n",
    "           total_time, total_time/60))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b0e8032d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>tu</th><th scope=col>gene.id</th><th scope=col>name</th><th scope=col>utr</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>  70006</td><td>  71585</td><td>1580</td><td>+</td><td>TU4 </td><td>ENSG00000186092</td><td>OR4F5  </td><td>UTR1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td> 944151</td><td> 944575</td><td> 425</td><td>+</td><td>TU17</td><td>ENSG00000187634</td><td>SAMD11 </td><td>UTR2</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td> 965189</td><td> 965719</td><td> 531</td><td>+</td><td>TU18</td><td>ENSG00000187961</td><td>KLHL17 </td><td>UTR3</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td> 974573</td><td> 975865</td><td>1293</td><td>+</td><td>TU19</td><td>ENSG00000187583</td><td>PLEKHN1</td><td>UTR4</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>1014476</td><td>1014540</td><td>  65</td><td>+</td><td>TU20</td><td>ENSG00000187608</td><td>ISG15  </td><td>UTR5</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>1054979</td><td>1056118</td><td>1140</td><td>+</td><td>TU21</td><td>ENSG00000188157</td><td>AGRN   </td><td>UTR6</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & seqnames & start & end & width & strand & tu & gene.id & name & utr\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 &   70006 &   71585 & 1580 & + & TU4  & ENSG00000186092 & OR4F5   & UTR1\\\\\n",
       "\t2 & chr1 &  944151 &  944575 &  425 & + & TU17 & ENSG00000187634 & SAMD11  & UTR2\\\\\n",
       "\t3 & chr1 &  965189 &  965719 &  531 & + & TU18 & ENSG00000187961 & KLHL17  & UTR3\\\\\n",
       "\t4 & chr1 &  974573 &  975865 & 1293 & + & TU19 & ENSG00000187583 & PLEKHN1 & UTR4\\\\\n",
       "\t5 & chr1 & 1014476 & 1014540 &   65 & + & TU20 & ENSG00000187608 & ISG15   & UTR5\\\\\n",
       "\t6 & chr1 & 1054979 & 1056118 & 1140 & + & TU21 & ENSG00000188157 & AGRN    & UTR6\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | tu &lt;chr&gt; | gene.id &lt;chr&gt; | name &lt;chr&gt; | utr &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 |   70006 |   71585 | 1580 | + | TU4  | ENSG00000186092 | OR4F5   | UTR1 |\n",
       "| 2 | chr1 |  944151 |  944575 |  425 | + | TU17 | ENSG00000187634 | SAMD11  | UTR2 |\n",
       "| 3 | chr1 |  965189 |  965719 |  531 | + | TU18 | ENSG00000187961 | KLHL17  | UTR3 |\n",
       "| 4 | chr1 |  974573 |  975865 | 1293 | + | TU19 | ENSG00000187583 | PLEKHN1 | UTR4 |\n",
       "| 5 | chr1 | 1014476 | 1014540 |   65 | + | TU20 | ENSG00000187608 | ISG15   | UTR5 |\n",
       "| 6 | chr1 | 1054979 | 1056118 | 1140 | + | TU21 | ENSG00000188157 | AGRN    | UTR6 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start   end     width strand tu   gene.id         name    utr \n",
       "1 chr1       70006   71585 1580  +      TU4  ENSG00000186092 OR4F5   UTR1\n",
       "2 chr1      944151  944575  425  +      TU17 ENSG00000187634 SAMD11  UTR2\n",
       "3 chr1      965189  965719  531  +      TU18 ENSG00000187961 KLHL17  UTR3\n",
       "4 chr1      974573  975865 1293  +      TU19 ENSG00000187583 PLEKHN1 UTR4\n",
       "5 chr1     1014476 1014540   65  +      TU20 ENSG00000187608 ISG15   UTR5\n",
       "6 chr1     1054979 1056118 1140  +      TU21 ENSG00000188157 AGRN    UTR6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# flank <- as.data.frame(genes$flank)\n",
    "df <- as.data.frame(genes$utr3)\n",
    "# df <- df %>%\n",
    "#  group_by(name) %>%\n",
    "#  arrange(name, ifelse(strand == '+', start, -start)) %>%\n",
    "#  mutate(most_downstream = row_number() == n()) %>%\n",
    "#  ungroup()\n",
    "# df$flank_start <- flank$start[match(df$name, flank$name)]\n",
    "# df$flank_end <- flank$end[match(df$name, flank$name)]\n",
    "# df$new_coord <- ifelse(df$strand == '+',\n",
    "#                        ifelse(df$most_downstream, df$end + 5000, df$end),\n",
    "#                        ifelse(df$most_downstream, df$start - 5000, df$start)) \n",
    "# df$end <- ifelse(df$strand == '+', df$new_coord, df$end)\n",
    "# df$start <- ifelse(df$strand == '-', df$new_coord, df$start)\n",
    "# df$width <- df$end - df$start+1\n",
    "head(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "67735813",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>name</th><th scope=col>score</th><th scope=col>strand</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>  70006</td><td>  71586</td><td>TU4_UTR1 </td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td> 944151</td><td> 944576</td><td>TU17_UTR2</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td> 965189</td><td> 965720</td><td>TU18_UTR3</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td> 974573</td><td> 975866</td><td>TU19_UTR4</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>1014476</td><td>1014541</td><td>TU20_UTR5</td><td>1</td><td>+</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>1054979</td><td>1056119</td><td>TU21_UTR6</td><td>1</td><td>+</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & chr & start & end & name & score & strand\\\\\n",
       "  & <fct> & <int> & <dbl> & <chr> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 &   70006 &   71586 & TU4\\_UTR1  & 1 & +\\\\\n",
       "\t2 & chr1 &  944151 &  944576 & TU17\\_UTR2 & 1 & +\\\\\n",
       "\t3 & chr1 &  965189 &  965720 & TU18\\_UTR3 & 1 & +\\\\\n",
       "\t4 & chr1 &  974573 &  975866 & TU19\\_UTR4 & 1 & +\\\\\n",
       "\t5 & chr1 & 1014476 & 1014541 & TU20\\_UTR5 & 1 & +\\\\\n",
       "\t6 & chr1 & 1054979 & 1056119 & TU21\\_UTR6 & 1 & +\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | chr &lt;fct&gt; | start &lt;int&gt; | end &lt;dbl&gt; | name &lt;chr&gt; | score &lt;dbl&gt; | strand &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 |   70006 |   71586 | TU4_UTR1  | 1 | + |\n",
       "| 2 | chr1 |  944151 |  944576 | TU17_UTR2 | 1 | + |\n",
       "| 3 | chr1 |  965189 |  965720 | TU18_UTR3 | 1 | + |\n",
       "| 4 | chr1 |  974573 |  975866 | TU19_UTR4 | 1 | + |\n",
       "| 5 | chr1 | 1014476 | 1014541 | TU20_UTR5 | 1 | + |\n",
       "| 6 | chr1 | 1054979 | 1056119 | TU21_UTR6 | 1 | + |\n",
       "\n"
      ],
      "text/plain": [
       "  chr  start   end     name      score strand\n",
       "1 chr1   70006   71586 TU4_UTR1  1     +     \n",
       "2 chr1  944151  944576 TU17_UTR2 1     +     \n",
       "3 chr1  965189  965720 TU18_UTR3 1     +     \n",
       "4 chr1  974573  975866 TU19_UTR4 1     +     \n",
       "5 chr1 1014476 1014541 TU20_UTR5 1     +     \n",
       "6 chr1 1054979 1056119 TU21_UTR6 1     +     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Sequences with length less than 8 are too short for SEA\n",
    "df <- df[df$width >= 8,]\n",
    "df$score <- 1\n",
    "df$tu <- paste0(df$tu,'_',df$utr)\n",
    "all_TU_disjoint_utr3 <- df %>% select(seqnames, start, end, tu, score, strand)\n",
    "colnames(all_TU_disjoint_utr3) <- c(\"chr\", \"start\", \"end\", \"name\", \"score\", \"strand\")\n",
    "all_TU_disjoint_utr3$end <- all_TU_disjoint_utr3$end + 1\n",
    "head(all_TU_disjoint_utr3)\n",
    "# write.table(all_TU_disjoint_utr3,paste0('all_TU_disjoint_utr3_5kbflank.bed'),sep='\\t',col.names=FALSE,row.names=FALSE,quote=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "97d7b9aa",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of TUs: 19002\n",
      "\n"
     ]
    }
   ],
   "source": [
    "num_TUs <- gsub('_UTR[0-9]+$', '', all_TU_disjoint_utr3$name) %>% unique() %>% length()\n",
    "message('Number of TUs: ', num_TUs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e3fe4770",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Number of UTR3s: 28581\n",
      "\n"
     ]
    }
   ],
   "source": [
    "message('Number of UTR3s: ', nrow(all_TU_disjoint_utr3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ed35076",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.5.1 (2025-06-13)\n",
       "Platform: aarch64-apple-darwin20\n",
       "Running under: macOS Ventura 13.0\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib \n",
       "LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1\n",
       "\n",
       "locale:\n",
       "[1] en_US/en_US/en_US/C/en_US/en_US\n",
       "\n",
       "time zone: America/New_York\n",
       "tzcode source: internal\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] pbapply_1.7-4        dplyr_1.1.4          goldmine_1.0        \n",
       " [4] ggplot2_4.0.0        stringr_1.5.2        data.table_1.17.8   \n",
       " [7] GenomicRanges_1.60.0 GenomeInfoDb_1.44.1  IRanges_2.42.0      \n",
       "[10] S4Vectors_0.46.0     BiocGenerics_0.54.0  generics_0.1.4      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] gtable_0.3.6            jsonlite_2.0.0          compiler_4.5.1         \n",
       " [4] crayon_1.5.3            tidyselect_1.2.1        IRdisplay_1.1          \n",
       " [7] parallel_4.5.1          systemfonts_1.2.3       scales_1.4.0           \n",
       "[10] textshaping_1.0.1       uuid_1.2-1              fastmap_1.2.0          \n",
       "[13] IRkernel_1.3.2          R6_2.6.1                XVector_0.48.0         \n",
       "[16] tibble_3.3.0            GenomeInfoDbData_1.2.14 pillar_1.11.1          \n",
       "[19] RColorBrewer_1.1-3      rlang_1.1.6             stringi_1.8.7          \n",
       "[22] repr_1.1.7              S7_0.2.0                cli_3.6.5              \n",
       "[25] withr_3.0.2             magrittr_2.0.4          digest_0.6.37          \n",
       "[28] grid_4.5.1              base64enc_0.1-3         pbdZMQ_0.3-14          \n",
       "[31] lifecycle_1.0.4         vctrs_0.6.5             evaluate_1.0.5         \n",
       "[34] glue_1.8.0              farver_2.1.2            ragg_1.4.0             \n",
       "[37] httr_1.4.7              pkgconfig_2.0.3         tools_4.5.1            \n",
       "[40] htmltools_0.5.8.1       UCSC.utils_1.4.0       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40bcc33e",
   "metadata": {},
   "source": [
    "### Extract UTR3 sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4faeefd1",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "!ref_genome=/Users/ninhle/Desktop/Research/bioinformatics_tools/ref_genomes/refdata-gex-GRCh38-2020-A/fasta/genome.fa\n",
    "!bed_input=all_TU_disjoint_utr3.bed\n",
    "!fa_output=all_TU_disjoint_utr3.fa\n",
    "!bedtools getfasta -fi $ref_genome -bed $bed_input -s -nameOnly -fo $fa_output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "598eb3c6",
   "metadata": {},
   "source": [
    "### Run SEA motif enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ef23b4b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "!docker run -v `pwd`:/home/meme/ memesuite/memesuite:latest sea --verbosity 1 --thresh 10.0 --align center --p /home/meme/all_TU_disjoint_utr3.fa --m /home/meme/meme_motif_results/raw_results/appMEME_5.5.517125209323241165022785_meme.txt -o sea_out_lgthn_all_TU_disjoint_utr3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9655866",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "!docker run -v `pwd`:/home/meme/ memesuite/memesuite:latest sea --verbosity 1 --thresh 10.0 --align center --p /home/meme/all_TU_disjoint_utr3_5kbflank.fa --m /home/meme/meme_motif_results/raw_results/appMEME_5.5.517125209323241165022785_meme.txt -o sea_out_lgthn_all_TU_disjoint_utr3_5kbflank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d10a565e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "!docker run -v `pwd`:/home/meme/ memesuite/memesuite:latest sea --verbosity 1 --thresh 10.0 --align center --p /home/meme/all_TU_disjoint_utr3_5kbflank.fa --m /home/meme/meme_motif_results/raw_results/appMEME_5.5.51712521030229-1466392474_meme.txt -o sea_out_shrtn_all_TU_disjoint_utr3_5kbflank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e54e34d4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "!docker run -v `pwd`:/home/meme/ memesuite/memesuite:latest sea --verbosity 1 --thresh 10.0 --align center --p /home/meme/all_TU_disjoint_utr3.fa --m /home/meme/meme_motif_results/raw_results/appMEME_5.5.51712521030229-1466392474_meme.txt -o sea_out_shrtn_all_TU_disjoint_utr3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
